{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cea562fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script type=\"text/javascript\" id=\"colomoto-setup-6428371\" class=\"to-be-removed\">\n",
       "        if (typeof Jupyter != 'undefined') {\n",
       "            \n",
       "function detect_import(cell, module) {\n",
       "    var code = cell.get_text();\n",
       "    code = code.replace(/\\\\\\n/g, \"\");\n",
       "    var lines = code.split(\"\\n\");\n",
       "    var r_simple = new RegExp(\"^(\"+module+\")$\");\n",
       "    var r_alias = new RegExp(\"^\"+module+\"\\\\s+as\\\\s+(\\\\w+)$\");\n",
       "    for (var i = 0; i < lines.length; ++i) {\n",
       "        if (/^import\\s/.test(lines[i])) {\n",
       "            code = lines[i].substr(7);\n",
       "            var parts = code.split(\",\")\n",
       "            for (var j = 0; j < parts.length; ++j) {\n",
       "                code = parts[j].trim();\n",
       "                var m = code.match(r_simple);\n",
       "                if (!m) {\n",
       "                    m = code.match(r_alias);\n",
       "                }\n",
       "                if (m) {\n",
       "                    return m[1];\n",
       "                }\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "    return module;\n",
       "}\n",
       "\n",
       "function colomoto_replace_call(cell, orig, dest, args, comment=false) {\n",
       "    var call_regexp = new RegExp(\"\\\\.\"+orig.replace(\".\",\"\\\\.\")+\"\\\\(\");\n",
       "    var call_replacer = new RegExp(\"\\\\.\"+orig.replace(\".\",\"\\\\.\")\n",
       "                    + \"\\\\(\\\\s*([^\\\\)]*)?\\\\)\");\n",
       "    var code = cell.get_text();\n",
       "    var lines = code.split(\"\\n\");\n",
       "    if (args) {\n",
       "        var strargs = \", \"+args.join(\", \");\n",
       "    } else {\n",
       "        var strargs = \"\";\n",
       "    }\n",
       "    for (var i = 0; i < lines.length; ++i) {\n",
       "        if (call_regexp.test(lines[i])) {\n",
       "            var code = \"\"\n",
       "            if (comment) {\n",
       "                code += \"#\"+lines[i]+\"\\n\";\n",
       "            }\n",
       "            code += lines[i].replace(call_replacer, \".\"+dest+\"($1\"+strargs+\")\")\n",
       "            lines[i] = code;\n",
       "        }\n",
       "    }\n",
       "    cell.set_text(code)\n",
       "}\n",
       "\n",
       "function colomoto_upload(Jupyter, ssid, input, py_callback_name, orig, dest) {\n",
       "\n",
       "    function callback(out_data) {\n",
       "        var cell_element = $(\"#\"+ssid).parents('.cell');\n",
       "        var cell_idx = Jupyter.notebook.get_cell_elements().index(cell_element);\n",
       "        var cell = Jupyter.notebook.get_cell(cell_idx);\n",
       "\n",
       "        var filename = out_data.content.text;\n",
       "\n",
       "        var code = cell.get_text();\n",
       "        code = code.replace(new RegExp(\"\\\\b\" + orig.replace('.', '\\\\.')\n",
       "                    + \"\\\\(\\\\s*((\\\\w+)=[^\\\\)]*)?\\\\)\"),\n",
       "                dest+\"(\\\"\"+filename+\"\\\",$1)\");\n",
       "        code = code.replace('\",)', '\")')\n",
       "        cell.set_text(code);\n",
       "\n",
       "        Jupyter.notebook.select(cell_idx);\n",
       "        Jupyter.notebook.execute_cell_and_select_below();\n",
       "    }\n",
       "\n",
       "    if (! (window.File && window.FileReader && window.FileList && window.Blob)) {\n",
       "        alert(\"Interactive file upload is not supported by your browser.\");\n",
       "        return;\n",
       "    }\n",
       "\n",
       "    input.disabled = true;\n",
       "    input.style.cursor = \"wait\";\n",
       "    input.parentElement.style.cursor = \"wait\";\n",
       "\n",
       "    var f = input.files[0];\n",
       "    var reader = new FileReader();\n",
       "    reader.onload = (function(f) {\n",
       "        return function (e) {\n",
       "            var obj = {\n",
       "                content: e.target.result,\n",
       "                name: f.name\n",
       "            };\n",
       "\n",
       "            //var pycb = py_callback_name+\"(\"+JSON.stringify(obj)+\")\"\n",
       "            // hack/workaround:\n",
       "            // it seems that Jupyter does not like very long lines\n",
       "            // so we split the data in chunks\n",
       "            var chunk_length = 100;\n",
       "            var pycb = \"__colomoto_upload_name = \" + JSON.stringify(obj.name) + \"\\n\";\n",
       "            pycb += \"__colomoto_upload_content = \\\\\\n\";\n",
       "            for (var i = 0; i < obj.content.length; i += chunk_length) {\n",
       "                pycb += \"\\\"\" + obj.content.substr(i, chunk_length)+\"\\\"\\\\\\n\"\n",
       "            }\n",
       "            pycb += \"\\n\"\n",
       "            pycb += py_callback_name+\"({'name':__colomoto_upload_name, 'content': __colomoto_upload_content})\\n\";\n",
       "            pycb += \"del __colomoto_upload_name, __colomoto_upload_content\"\n",
       "\n",
       "            IPython.notebook.kernel.execute(pycb, {iopub: {output: callback}});\n",
       "        };\n",
       "    })(f);\n",
       "    reader.readAsDataURL(f);\n",
       "}\n",
       "\n",
       "function resolve_function(tool_api, funcname) {\n",
       "    if (tool_api.hasOwnProperty(funcname)) {\n",
       "        return tool_api[funcname];\n",
       "    } else {\n",
       "        return window[funcname];\n",
       "    }\n",
       "}\n",
       "\n",
       "function colomoto_extension(Jupyter, ssid, name, menu, toolbar, tool_api) {\n",
       "\n",
       "    function insert_snippet_code(snippet) {\n",
       "        var cell = Jupyter.notebook.get_selected_cell();\n",
       "        Jupyter.notebook.edit_mode();\n",
       "        cell.code_mirror.replaceSelection(snippet, 'around');\n",
       "        //cell.focus_editor();\n",
       "    }\n",
       "\n",
       "    /**\n",
       "        from https://github.com/moble/jupyter_boilerplate/blob/master/main.js\n",
       "    */\n",
       "    function callback_insert_snippet (evt) {\n",
       "        // this (or event.currentTarget, see below) always refers to the DOM\n",
       "        // element the listener was attached to - see\n",
       "        // http://stackoverflow.com/questions/12077859\n",
       "        insert_snippet_code($(evt.currentTarget).data('snippet-code'));\n",
       "    }\n",
       "    function build_menu_element (menu_item_spec, direction) {\n",
       "        // Create the menu item html element\n",
       "        var element = $('<li/>');\n",
       "\n",
       "        if (typeof menu_item_spec == 'string') {\n",
       "            if (menu_item_spec != '---') {\n",
       "                return element.html(menu_item_spec)\n",
       "                        .addClass('ui-state-disabled')\n",
       "                        .attr({\"style\": \"padding:2px .4em\"})\n",
       "                       ;\n",
       "            }\n",
       "            return element.addClass('divider');\n",
       "        }\n",
       "\n",
       "        var a = $('<a/>')\n",
       "            .attr('href', '#')\n",
       "            .html(menu_item_spec.name)\n",
       "            .appendTo(element);\n",
       "        if (menu_item_spec.hasOwnProperty('snippet')) {\n",
       "            var snippet = menu_item_spec.snippet;\n",
       "            if (typeof snippet == 'string' || snippet instanceof String) {\n",
       "                snippet = [snippet];\n",
       "            }\n",
       "            a.attr({\n",
       "                'title' : \"\", // Do not remove this, even though it's empty!\n",
       "                'data-snippet-code' : snippet.join('\\n'),\n",
       "            })\n",
       "            .on('click', callback_insert_snippet)\n",
       "            .addClass('snippet');\n",
       "        }\n",
       "        else if (menu_item_spec.hasOwnProperty('internal-link')) {\n",
       "            a.attr('href', menu_item_spec['internal-link']);\n",
       "        }\n",
       "        else if (menu_item_spec.hasOwnProperty('external-link')) {\n",
       "            a.empty();\n",
       "            a.attr('href', menu_item_spec['external-link']);\n",
       "            a.attr({\n",
       "                'target' : '_blank',\n",
       "                'title' : 'Opens in a new window',\n",
       "            });\n",
       "            $('<i class=\"fa fa-external-link menu-icon pull-right\"/>').appendTo(a);\n",
       "            $('<span/>').html(menu_item_spec.name).appendTo(a);\n",
       "        }\n",
       "\n",
       "        if (menu_item_spec.hasOwnProperty('sub-menu')) {\n",
       "            element\n",
       "                .addClass('dropdown-submenu')\n",
       "                .toggleClass('dropdown-submenu-left', direction === 'left');\n",
       "            var sub_element = $('<ul class=\"dropdown-menu\"/>')\n",
       "                .toggleClass('dropdown-menu-compact', menu_item_spec.overlay === true) // For space-saving menus\n",
       "                .appendTo(element);\n",
       "\n",
       "            var new_direction = (menu_item_spec['sub-menu-direction'] === 'left') ? 'left' : 'right';\n",
       "            for (var j=0; j<menu_item_spec['sub-menu'].length; ++j) {\n",
       "                var sub_menu_item_spec = build_menu_element(menu_item_spec['sub-menu'][j], new_direction);\n",
       "                if(sub_menu_item_spec !== null) {\n",
       "                    sub_menu_item_spec.appendTo(sub_element);\n",
       "                }\n",
       "            }\n",
       "        }\n",
       "\n",
       "        return element;\n",
       "    }\n",
       "\n",
       "    function menu_setup (menu_item_specs, sibling, insert_before_sibling) {\n",
       "        for (var i=0; i<menu_item_specs.length; ++i) {\n",
       "            var menu_item_spec;\n",
       "            if (insert_before_sibling) {\n",
       "                menu_item_spec = menu_item_specs[i];\n",
       "            } else {\n",
       "                menu_item_spec = menu_item_specs[menu_item_specs.length-1-i];\n",
       "            }\n",
       "            var direction = (menu_item_spec['menu-direction'] == 'left') ? 'left' : 'right';\n",
       "            var menu_element = build_menu_element(menu_item_spec, direction);\n",
       "            // We need special properties if this item is in the navbar\n",
       "            if ($(sibling).parent().is('ul.nav.navbar-nav')) {\n",
       "                menu_element\n",
       "                    .addClass('dropdown')\n",
       "                    .removeClass('dropdown-submenu dropdown-submenu-left');\n",
       "                menu_element.children('a')\n",
       "                    .addClass('dropdown-toggle')\n",
       "                    .attr({\n",
       "                        'id': name+'_menu',\n",
       "                        'data-toggle' : 'dropdown',\n",
       "                        'aria-expanded' : 'false'\n",
       "                    });\n",
       "            }\n",
       "\n",
       "            // Insert the menu element into DOM\n",
       "            menu_element[insert_before_sibling ? 'insertBefore': 'insertAfter'](sibling);\n",
       "        }\n",
       "    }\n",
       "    /** end from */\n",
       "\n",
       "\n",
       "    function self_cleanup() {\n",
       "        var cell_element = $(\"script[class='to-be-removed']\").parents('.cell');\n",
       "        var cell_idx = Jupyter.notebook.get_cell_elements().index(cell_element);\n",
       "        var cell = Jupyter.notebook.get_cell(cell_idx);\n",
       "        var to_remove = -1;\n",
       "        for (var i = 0; i < cell.output_area.outputs.length; ++i) {\n",
       "            var oa = cell.output_area.outputs[i];\n",
       "            if (oa.output_type == \"display_data\"\n",
       "                && typeof oa.data[\"text/html\"] != 'undefined'\n",
       "                && oa.data[\"text/html\"].indexOf(' class=\"to-be-removed\"') >= 0) {\n",
       "                to_remove = i;\n",
       "                break;\n",
       "            }\n",
       "        }\n",
       "        if (to_remove == -1) {\n",
       "            console.log(\"cannot find toberemoved\");\n",
       "        } else {\n",
       "            cell.output_area.outputs.splice(to_remove, 1);\n",
       "        }\n",
       "    }\n",
       "\n",
       "    function toolbar_setup(actions) {\n",
       "        var buttons = [];\n",
       "        for (var i = 0; i < actions.length; ++i) {\n",
       "            var setup = actions[i].setup;\n",
       "            if (typeof setup.handler == 'string') {\n",
       "                setup.handler = resolve_function(tool_api, setup.handler);\n",
       "            }\n",
       "            buttons.push(Jupyter.actions.register(actions[i].setup,\n",
       "                actions[i].name, name));\n",
       "        }\n",
       "        $(\"#\"+name+\"-toolbar\").remove();\n",
       "        Jupyter.toolbar.add_buttons_group(buttons, name+\"-toolbar\");\n",
       "    }\n",
       "\n",
       "    function replace_menu_snippets(menu_spec, orig, dest) {\n",
       "        if (menu_spec.hasOwnProperty(\"snippet\")) {\n",
       "            var snippet = menu_spec.snippet;\n",
       "            if (typeof snippet == \"string\" || snippet instanceof String) {\n",
       "                menu_spec[\"snippet\"] = snippet.replace(orig, dest);\n",
       "            } else {\n",
       "                for (var i = 0; i < snippet.length; ++i) {\n",
       "                    menu_spec[\"snippet\"][i] = snippet[i].replace(orig, dest);\n",
       "                }\n",
       "            }\n",
       "        }\n",
       "        if (menu_spec.hasOwnProperty(\"sub-menu\")) {\n",
       "            for (var i = 0; i < menu_spec[\"sub-menu\"].length; ++i) {\n",
       "                replace_menu_snippets(menu_spec[\"sub-menu\"][i], orig, dest);\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "\n",
       "    function load_ipython_extension() {\n",
       "\n",
       "        var mycellelt = $(\"#\"+ssid).parents('.cell');\n",
       "        var myidx = Jupyter.notebook.get_cell_elements().index(mycellelt);\n",
       "        var import_cell = Jupyter.notebook.get_cell(myidx);\n",
       "\n",
       "        var alias = detect_import(import_cell, name);\n",
       "        tool_api.module_alias = alias;\n",
       "        if (alias && alias != name) {\n",
       "            var orig = new RegExp(\"\\\\b\"+name+\"\\\\b\", \"g\");\n",
       "            replace_menu_snippets(menu, orig, alias);\n",
       "        }\n",
       "\n",
       "        if (toolbar) {\n",
       "            toolbar_setup(toolbar);\n",
       "        }\n",
       "\n",
       "        $(\"#\"+name+\"_menu\").parent().remove();\n",
       "        if (menu) {\n",
       "            menu_setup([menu], $(\"#help_menu\").parent(), true);\n",
       "        }\n",
       "\n",
       "        if (tool_api.hasOwnProperty(\"post_install_callback\")) {\n",
       "            tool_api.post_install_callback();\n",
       "        }\n",
       "\n",
       "        setTimeout(self_cleanup, 5000);\n",
       "    };\n",
       "\n",
       "    load_ipython_extension();\n",
       "}\n",
       "\n",
       "function resolve_toolbar_handlers(tool_api, toolbar_spec) {\n",
       "    for (var i = 0; i < toolbar_spec.length; ++i) {\n",
       "        func = resolve_function(tool_api, toolbar_spec[i][\"setup\"][\"handler\"]);\n",
       "        toolbar_spec[i][\"setup\"][\"handler\"] = func;\n",
       "    }\n",
       "    return toolbar_spec\n",
       "}\n",
       "\n",
       "\n",
       "    var minibn_jsapi = {  };\n",
       "    colomoto_extension(Jupyter, \"colomoto-setup-6428371\", \"minibn\", null, null, minibn_jsapi);\n",
       "     }</script>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<script type=\"text/javascript\" id=\"colomoto-setup-515203\" class=\"to-be-removed\">\n",
       "        if (typeof Jupyter != 'undefined') {\n",
       "            \n",
       "    var mymodule_jsapi = {  };\n",
       "    colomoto_extension(Jupyter, \"colomoto-setup-515203\", \"mymodule\", {\"name\": \"MaBoSS\", \"sub-menu\": [{\"name\": \"Load MaBoSS file\", \"snippet\": [\"masim = maboss.load(\\\"filename.bnd\\\", \\\"filename.cfg\\\")\"]}, {\"name\": \"Network\", \"sub-menu\": [{\"name\": \"Set istate\", \"snippet\": [\"maboss.wg_set_istate(masim)\"]}, {\"name\": \"Get initial state\", \"snippet\": [\"masim.get_initial_state()\"]}]}, {\"name\": \"Simulation\", \"sub-menu\": [{\"name\": \"Create mutant\", \"snippet\": [\"maboss.wg_make_mutant(masim)\"]}, {\"name\": \"Run\", \"snippet\": [\"simres = masim.run()\"]}, {\"name\": \"Set initial states\", \"snippet\": [\"masim.network.set_istate([\\\"mygenelist\\\"],{})\"]}, {\"name\": \"Set output\", \"snippet\": [\"maboss.wg_set_output(masim)\"]}]}, {\"name\": \"Results\", \"sub-menu\": [{\"name\": \"Save results\", \"snippet\": [\"simres.save(\\\"filename\\\")\"]}, {\"name\": \"Plot piechart\", \"snippet\": [\"simres.plot_piechart()\"]}, {\"name\": \"Plot trajectory\", \"snippet\": [\"simres.plot_trajectory()\"]}]}, {\"name\": \"Conversion\", \"sub-menu\": [{\"name\": \"Convert to biolqm\", \"snippet\": [\"biolqm_model = maboss.to_biolqm(masim)\"]}, {\"name\": \"Convert to minibn\", \"snippet\": [\"minibn_model = maboss.to_minibn(masim)\"]}]}, \"---\", {\"name\": \"Documentation\", \"external-link\": \"http://pymaboss.readthedocs.io\"}]}, null, mymodule_jsapi);\n",
       "     }</script><style type=\"text/css\">\n",
       "    #mymodule_menu {\n",
       "        color: green;\n",
       "        background-color: white;\n",
       "    }\n",
       "    </style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<script type=\"text/javascript\" id=\"colomoto-setup-3134390\" class=\"to-be-removed\">\n",
       "        if (typeof Jupyter != 'undefined') {\n",
       "            \n",
       "    var biolqm_jsapi = { action_upload_model: function() {\n",
       "        var cell = Jupyter.notebook.get_selected_cell();\n",
       "        cell.set_text('lqm = '+biolqm_jsapi.module_alias+'.upload()');\n",
       "        cell.focus_editor();\n",
       "    } };\n",
       "    colomoto_extension(Jupyter, \"colomoto-setup-3134390\", \"biolqm\", {\"name\": \"bioLQM\", \"sub-menu\": [{\"name\": \"Upload model\", \"snippet\": [\"lqm = biolqm.upload()\"]}, {\"name\": \"Load model\", \"snippet\": [\"lqm = biolqm.load(\\\"model.sbml\\\")\"]}, \"---\", {\"name\": \"Export to file\", \"sub-menu\": [{\"name\": \"SBML-qual v1.0 (.sbml)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.sbml\\\", \\\"sbml\\\")\"]}, \"Functions formats\", {\"name\": \"BoolNet (.bnet)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.bnet\\\", \\\"bnet\\\")\"]}, {\"name\": \"BooleanNet (.booleannet)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.booleannet\\\", \\\"booleannet\\\")\"]}, {\"name\": \"BoolSim (.boolsim)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.boolsim\\\", \\\"boolsim\\\")\"]}, {\"name\": \"Raw Boolean functions (.boolfunctions)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.boolfunctions\\\", \\\"boolfunctions\\\")\"]}, {\"name\": \"Truth table (.tt)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.tt\\\", \\\"tt\\\")\"]}, \"Petri net formats\", {\"name\": \"APNN (.apnn)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.apnn\\\", \\\"apnn\\\")\"]}, {\"name\": \"INA (.ina)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.ina\\\", \\\"ina\\\")\"]}, {\"name\": \"PNML (.pnml)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.pnml\\\", \\\"pnml\\\")\"]}, \"Dedicated formats\", {\"name\": \"GINML (.ginml)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.ginml\\\", \\\"ginml\\\")\"]}, {\"name\": \"GNA non-xml (.gna)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.gna\\\", \\\"gna\\\")\"]}, {\"name\": \"MaBoSS (.bnd)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.bnd\\\", \\\"bnd\\\")\"]}, {\"name\": \"Pint (.an)\", \"snippet\": [\"biolqm.save(lqm, \\\"mymodel.an\\\", \\\"an\\\")\"]}]}, {\"name\": \"Convert to tool\", \"sub-menu\": [{\"name\": \"GINsim\", \"snippet\": [\"lrg = biolqm.to_ginsim(lqm)\"]}, {\"name\": \"MaBoSS\", \"snippet\": [\"masim = biolqm.to_maboss(lqm)\"]}, {\"name\": \"Pint\", \"snippet\": [\"an = biolqm.to_pint(lqm)\"]}]}, \"---\", {\"name\": \"Compute fixpoints\", \"snippet\": [\"fps = biolqm.fixpoints(lqm)\"]}, {\"name\": \"Compute trap spaces\", \"snippet\": [\"traps = biolqm.trapspace(lqm)\"]}, \"---\", {\"name\": \"Model modifications\", \"sub-menu\": [{\"name\": \"Perturbation\", \"snippet\": [\"lqm_mod = biolqm.perturbation(lqm, \\\"node%0\\\")\"]}, {\"name\": \"Booleanization\", \"snippet\": [\"lqm_bool = biolqm.booleanize(lqm)\"]}, {\"name\": \"Reduction\", \"snippet\": [\"lqm_red = biolqm.reduce(lqm, \\\"fixed,output,duplicate\\\")\"]}, {\"name\": \"Reversal\", \"snippet\": [\"lqm_rev = biolqm.reverse(lqm)\"]}, {\"name\": \"Sanitize\", \"snippet\": [\"lqm_san = biolqm.sanitize(lqm)\"]}]}, \"---\", {\"name\": \"Documentation\", \"external-link\": \"http://ginsim.naldi.info/biolqm/site/doc/index.html\"}]}, [{\"name\": \"upload\", \"setup\": {\"icon\": \"fa-upload\", \"help\": \"Upload model\", \"handler\": \"action_upload_model\"}}], biolqm_jsapi);\n",
       "     }</script><style type=\"text/css\">\n",
       "    #biolqm_menu {\n",
       "        color: #00007f;\n",
       "        background-color: white;\n",
       "    }\n",
       "    \n",
       "    div#biolqm-toolbar:before {\n",
       "        content: \"bioLQM\";\n",
       "        float: left;\n",
       "        margin-left: 1em;\n",
       "        margin-right: 1em;\n",
       "        margin-top: 0.2em;\n",
       "    }\n",
       "    div#biolqm-toolbar,\n",
       "    div#biolqm-toolbar button {\n",
       "        color: #00007f;\n",
       "    }\n",
       "    </style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<script type=\"text/javascript\" id=\"colomoto-setup-2822438\" class=\"to-be-removed\">\n",
       "        if (typeof Jupyter != 'undefined') {\n",
       "            \n",
       "    var ginsim_jsapi = { action_upload_model: function() {\n",
       "        var cell = Jupyter.notebook.get_selected_cell();\n",
       "        cell.set_text('lrg = '+ginsim_jsapi.module_alias+'.upload()');\n",
       "        cell.focus_editor();\n",
       "    } };\n",
       "    colomoto_extension(Jupyter, \"colomoto-setup-2822438\", \"ginsim\", {\"name\": \"GINsim\", \"sub-menu\": [{\"name\": \"Upload model\", \"snippet\": [\"lrg = ginsim.upload()\"]}, {\"name\": \"Load model\", \"snippet\": [\"lrg = ginsim.load(\\\"model.zginml\\\")\"]}, \"---\", {\"name\": \"Export to file\", \"sub-menu\": [{\"name\": \"MaBoSS (.bnd, .cfg)\", \"snippet\": [\"ginsim.service(\\\"maboss\\\").export(lrg, \\\"mymodel.bnd\\\")\"]}]}, {\"name\": \"Convert to tool\", \"sub-menu\": [{\"name\": \"bioLQM\", \"snippet\": [\"lqm = ginsim.to_biolqm(lrg)\"]}, {\"name\": \"MaBoSS\", \"snippet\": [\"masim = ginsim.to_maboss(lrg)\"]}, {\"name\": \"NuSMV\", \"snippet\": [\"smv = ginsim.to_nusmv(lrg)\"]}, {\"name\": \"Pint\", \"snippet\": [\"an = ginsim.to_pint(lrg)\"]}]}, \"---\", {\"name\": \" Display regulatory graph\", \"snippet\": [\"ginsim.show(lrg)\"]}]}, [{\"name\": \"upload\", \"setup\": {\"icon\": \"fa-upload\", \"help\": \"Upload model\", \"handler\": \"action_upload_model\"}}], ginsim_jsapi);\n",
       "     }</script><style type=\"text/css\">\n",
       "    #ginsim_menu {\n",
       "        color: blue;\n",
       "        background-color: white;\n",
       "    }\n",
       "    \n",
       "    div#ginsim-toolbar:before {\n",
       "        content: \"GINsim\";\n",
       "        float: left;\n",
       "        margin-left: 1em;\n",
       "        margin-right: 1em;\n",
       "        margin-top: 0.2em;\n",
       "    }\n",
       "    div#ginsim-toolbar,\n",
       "    div#ginsim-toolbar button {\n",
       "        color: blue;\n",
       "    }\n",
       "    </style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import maboss\n",
    "import ginsim\n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "\n",
    "\n",
    "from identification_patients.get_patients_sens_res import get_patients\n",
    "\n",
    "from create_generic_models.create_generic_patients_cfgs import create_generic_patients_cfgs_bnds\n",
    "\n",
    "from create_generic_models.update_phenotypes_generic_models import generic_models_update_phenotypes\n",
    "from pre_process_data.pre_process_genes import create_table_rna_seq_patients, process_genes\n",
    "\n",
    "from create_person_models.tailor_cfgs_patients_gene import personalized_patients_genes_cfgs\n",
    "from create_person_models.tailor_bnd_mutations import personalized_patients_mutations_bnds\n",
    "from create_person_models.tailor_bnd_tsg_onco_mutations import tailor_bnd_tsg_onco_mut\n",
    "from create_person_models.tailor_bnd_cnv import tailor_bnd_cnv_cm\n",
    "\n",
    "\n",
    "from MaBoSS_simulation.MaBoSS_phenotype_distribution import compute_phenotypes_distribution, compute_mean_patients\n",
    "from pre_process_data.identify_mutations_patients import identif_mutations_kras_egfr\n",
    "\n",
    "# # from pre_process_profiles_table_data_lung import create_genes_patients\n",
    "\n",
    "from stats.stats_proba import compute_mannwhitneyu_test_means\n",
    "from results_MaBoSS_visualization.boxplot_phenotype import create_boxplot\n",
    "from results_MaBoSS_visualization.create_phenotypes_patients_table import vizualise_table_phenotype_condition\n",
    "from results_MaBoSS_visualization.patients_ids_phenotype_table import create_table_patients_phenotypes\n",
    "from gene_enrichment.genes_signature import compute_genes_mean_signature\n",
    "from pre_process_data.preprocess_mutations import preprocess_mutations_ds\n",
    "from pre_process_data.pre_process_cnv import preprocess_cnv\n",
    "\n",
    "\n",
    "\n",
    "# issues when compute phenotype proba > check if related to the phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "251ede88",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/romane/repos/icr/Attractor-Resistance\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "print(os.getcwd())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba96ba2f",
   "metadata": {},
   "source": [
    "Step 0: Import data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "543de951",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/83/pqwsm5cn7vscbq8l3fnf30qr0000gn/T/ipykernel_27224/4278837298.py:15: DtypeWarning: Columns (22,50,56,57,58,59,61) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  mutations_data_depseq = pd.read_csv('data/depmap_data/OmicsSomaticMutations.csv')\n"
     ]
    }
   ],
   "source": [
    "# Import Data\n",
    "\n",
    "annotations_models = pd.read_csv('data/model_list_20250407.csv')\n",
    "mutations_data = pd.read_csv('data/mutations_all_20250318.csv')\n",
    "drug_data = pd.read_csv('data/drug_sensitivity.csv')\n",
    "\n",
    "montagud_data = (\n",
    "    pd.read_csv('data/Montagud_inter_nodes_data.csv', header=1)\n",
    "    .loc[:, ['Target node', 'Interaction type', 'Source']])\n",
    "rna_seq_data = pd.read_csv('data/rnaseq_merged/rnaseq_merged_20250117.csv')\n",
    "cnv_data = pd.read_csv('data/cellmodel_data/cnv_summary_20250207.csv')\n",
    "\n",
    "\n",
    "models_depmap_annotation = pd.read_csv('data/cellmodel_data/model_list_20250423.csv')\n",
    "mutations_data_depseq = pd.read_csv('data/depmap_data/OmicsSomaticMutations.csv')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7c8a8476",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chose folder where we want all the personalized boolean models and associated results saved \n",
    "drug_interest = 'Refametinib' #Pictilisib, 'Avagacestat' AZD8931 \n",
    "tissue_interest = 'PAN_CANCER'\n",
    "tissue_remove = 'Haematopoietic and Lymphoid'\n",
    "\n",
    "folder_pers_models=f'models/personalized_boolean_{drug_interest}_{tissue_interest}'\n",
    "folder_result = f'results/{drug_interest}_{tissue_interest}'\n",
    "folder_result_resistant = f'{folder_result}/resistant_results/only_gene_expression/single_input_on/phenotype_distribution_patients'\n",
    "folder_result_sensitive = f'{folder_result}/sensitive_results/only_gene_expression/single_input_on/phenotype_distribution_patients'\n",
    "folder_result_res_sen = f'{folder_result}/sensitive_resistant_results/genes_diff_expressed'\n",
    "\n",
    "\n",
    "os.makedirs(folder_result_resistant, exist_ok=True)\n",
    "os.makedirs(folder_result_sensitive, exist_ok=True)\n",
    "os.makedirs(folder_result_res_sen, exist_ok=True)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Output directories\n",
    "output_dir_resistant = f'{folder_pers_models}/resistant_patient/generic_models'\n",
    "output_dir_sec_resistant = f'{folder_pers_models}/resistant_patient/personalized_boolean_modified/models_gene_expression'\n",
    "output_dir_sensitive = f'{folder_pers_models}/sensitive_patient/generic_models'\n",
    "output_dir_sec_sensitive = f'{folder_pers_models}/sensitive_patient/personalized_boolean_modified/models_gene_expression'\n",
    "bnd_dir_res = f'{folder_pers_models}/resistant_patient/personalized_boolean_modified/models_gene_expression'\n",
    "bnd_dir_sens = f'{folder_pers_models}/sensitive_patient/personalized_boolean_modified/models_gene_expression'\n",
    "\n",
    "os.makedirs(bnd_dir_sens, exist_ok=True)\n",
    "os.makedirs(bnd_dir_res, exist_ok=True)\n",
    "\n",
    "os.makedirs(output_dir_sec_resistant, exist_ok=True)\n",
    "os.makedirs(output_dir_sec_sensitive, exist_ok=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "631374cd",
   "metadata": {},
   "source": [
    "Step 1: Pre-process data- Montagud nodes, rna-seq (filtered with patients IDs and Montagud nodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a6a0b59a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pre-process Montagud nodes\n",
    "\n",
    "# Keep only the nodes of montagud not specific to Prostate \n",
    "montagud_nodes = list(set(montagud_data['Target node'].tolist() + montagud_data['Source'].tolist()))\n",
    "montagud_nodes = [node for node in montagud_nodes if node != '0/1']\n",
    "montagud_nodes = [node.upper() for node in montagud_nodes if isinstance(node, str)]\n",
    "montagud_nodes.append('KRAS')\n",
    "to_remove = ['RAS', 'FUSED_EVENT', 'NKX3_1', 'SPOP', 'AR_ERG']\n",
    "\n",
    "montagud_nodes = [node for node in montagud_nodes if node not in to_remove]\n",
    "montagud_nodes = list(set(montagud_nodes))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5db46ca",
   "metadata": {},
   "source": [
    "Step 1: Check what drug is the best to keep (the one with most resistant and sensitive)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a0562c6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# results = {}\n",
    "# drug_interests = drug_data['DRUG_NAME'].unique().tolist()\n",
    "\n",
    "# for drug in drug_interests:\n",
    "#     top_resistant_ids, top_sensitive_ids, drug_data_filtered = get_patients(500, drug_data, annotations_models, drug, tissue_remove)\n",
    "#     patients_ids = top_resistant_ids + top_sensitive_ids\n",
    "\n",
    "#     drug_tissue_filtered = drug_data_filtered[drug_data_filtered['SANGER_MODEL_ID'].isin(patients_ids)]    \n",
    "#     results[drug] = {\n",
    "#          \"name\": drug,\n",
    "#          \"<-1.5\": float((drug_tissue_filtered['Z_SCORE'] < -1.5).sum()),\n",
    "#          \">1.5\": float((drug_tissue_filtered['Z_SCORE'] > 1.5).sum()),\n",
    "#          \"mean\": float(drug_tissue_filtered['Z_SCORE'].mean()),\n",
    "#          \"std\": float(drug_tissue_filtered['Z_SCORE'].std()),\n",
    "#          \"abs_zscore\": float(drug_tissue_filtered['Z_SCORE'].abs().mean()),\n",
    "#      }\n",
    "\n",
    "# pd_results = pd.DataFrame(results.values())\n",
    "# pd_results.to_csv(f'{folder_result}/drug_analysis.csv', index=False)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66844528",
   "metadata": {},
   "source": [
    "Step 2: Select cancer and drug of interest (tissue_interest, drug_interest)\n",
    "Get the 100 most sensitive patient\n",
    "Pre-process data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1c471ff9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/romane/repos/icr/Attractor-Resistance/identification_patients/get_patients_sens_res.py:37: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  annotations_models_filtered.rename(\n"
     ]
    }
   ],
   "source": [
    "# Get patients IDs\n",
    "\n",
    "# changed the drug name to Pictilisib !!!!\n",
    "\n",
    "# top_resistant_ids, top_sensitive_ids= get_patients(drug_data, annotations_models, drug_interest, tissue_interest)\n",
    "top_resistant_ids, top_sensitive_ids, drug_tissue_data= get_patients(20, drug_data, annotations_models, drug_interest, tissue_interest = None, tissue_remove = tissue_remove)\n",
    "patients_ids = top_sensitive_ids + top_resistant_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8a9c1e8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/romane/repos/icr/Attractor-Resistance/pre_process_data/pre_process_genes.py:37: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  rna_seq_data.rename(columns={\"gene_symbol\": \"gene_name\"}, inplace=True)\n"
     ]
    }
   ],
   "source": [
    "# Pre-process RNA seq data (keep only montagud_nodes and patients ids) and create table for high and low expression\n",
    "rna_seq_data_filtered = process_genes(patients_ids, montagud_nodes, rna_seq_data)\n",
    "table_rna_seq_patients = create_table_rna_seq_patients(rna_seq_data_filtered)\n",
    "\n",
    "# check how many genes info we have the selected patients\n",
    "# print(rna_seq_data.head())\n",
    "# rna_seq_data = rna_seq_data[rna_seq_data['model_id'].isin(patients_ids)]\n",
    "# rna_seq_data = rna_seq_data[rna_seq_data['gene_symbol'].isin(montagud_nodes)]\n",
    "# print(rna_seq_data)\n",
    "# print(len(rna_seq_data['model_id'].unique().tolist()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6ff5c86b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                VariantType  \\\n",
      "cell model id                 \n",
      "SIDM00920          deletion   \n",
      "SIDM00506               SNV   \n",
      "SIDM00508          deletion   \n",
      "SIDM00292          deletion   \n",
      "SIDM00506               SNV   \n",
      "SIDM00524               SNV   \n",
      "SIDM00870               SNV   \n",
      "SIDM00795               SNV   \n",
      "SIDM00920               SNV   \n",
      "SIDM00711               SNV   \n",
      "SIDM00175         insertion   \n",
      "SIDM00183          deletion   \n",
      "SIDM01120               SNV   \n",
      "SIDM01120          deletion   \n",
      "SIDM00255               SNV   \n",
      "SIDM00183      substitution   \n",
      "\n",
      "                                                     VariantInfo HugoSymbol  \\\n",
      "cell model id                                                                 \n",
      "SIDM00920                                     frameshift_variant       PTEN   \n",
      "SIDM00506                                            stop_gained      BRCA2   \n",
      "SIDM00508                                     frameshift_variant        RB1   \n",
      "SIDM00292      splice_acceptor_variant&coding_sequence_varian...        RB1   \n",
      "SIDM00506                                splice_acceptor_variant        RB1   \n",
      "SIDM00524                                            stop_gained        RB1   \n",
      "SIDM00870                                splice_acceptor_variant        NF1   \n",
      "SIDM00795                                splice_acceptor_variant      EP300   \n",
      "SIDM00920                                            stop_gained        VHL   \n",
      "SIDM00711                                splice_acceptor_variant      MED12   \n",
      "SIDM00175                                     frameshift_variant       PTEN   \n",
      "SIDM00183                                     frameshift_variant        RB1   \n",
      "SIDM01120                                            stop_gained        RB1   \n",
      "SIDM01120                                     frameshift_variant      BRCA1   \n",
      "SIDM00255                                            stop_gained        ATR   \n",
      "SIDM00183                                            stop_gained      PTCH1   \n",
      "\n",
      "               OncogeneHighImpact  TumorSuppressorHighImpact   depmap id  \n",
      "cell model id                                                             \n",
      "SIDM00920                   False                       True  ACH-000783  \n",
      "SIDM00506                   False                       True  ACH-000508  \n",
      "SIDM00508                   False                       True  ACH-001489  \n",
      "SIDM00292                   False                       True  ACH-002077  \n",
      "SIDM00506                   False                       True  ACH-000508  \n",
      "SIDM00524                   False                       True  ACH-000703  \n",
      "SIDM00870                   False                       True  ACH-000725  \n",
      "SIDM00795                   False                       True  ACH-000219  \n",
      "SIDM00920                   False                       True  ACH-000783  \n",
      "SIDM00711                    True                       True  ACH-000727  \n",
      "SIDM00175                   False                       True  ACH-002096  \n",
      "SIDM00183                   False                       True  ACH-002154  \n",
      "SIDM01120                   False                       True  ACH-002174  \n",
      "SIDM01120                   False                       True  ACH-002174  \n",
      "SIDM00255                   False                       True  ACH-002282  \n",
      "SIDM00183                   False                       True  ACH-002154  \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/romane/repos/icr/Attractor-Resistance/pre_process_data/preprocess_mutations.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  models_annotation.rename(columns={'model_id': 'cell model id'}, inplace=True)\n",
      "/Users/romane/repos/icr/Attractor-Resistance/pre_process_data/preprocess_mutations.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  models_annotation.rename(columns={'BROAD_ID': 'depmap id'}, inplace=True)\n"
     ]
    }
   ],
   "source": [
    "mutations_data_filtered_combined = preprocess_mutations_ds(patients_ids, models_depmap_annotation, mutations_data_depseq,montagud_nodes)\n",
    "print(mutations_data_filtered_combined)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "abfece30",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         model_id gene_symbol  total_copy_number cn_category\n",
      "2313    SIDM00003        BCL2                2.0        Loss\n",
      "2502    SIDM00003         SMO                5.0        Gain\n",
      "2570    SIDM00003         JUN                4.0        Gain\n",
      "2591    SIDM00003        EZH2                5.0        Gain\n",
      "2594    SIDM00003         ERG                4.0        Gain\n",
      "...           ...         ...                ...         ...\n",
      "875480  SIDM01242         VHL                2.0        Loss\n",
      "875691  SIDM01242       FGFR3                2.0        Loss\n",
      "875825  SIDM01242       MED12                5.0        Gain\n",
      "875947  SIDM01242       FOXA1                2.0        Loss\n",
      "875957  SIDM01242        BCL2                1.0        Loss\n",
      "\n",
      "[468 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "# preprocess cnv\n",
    "cnv_data_filtered = preprocess_cnv(cnv_data, montagud_nodes,patients_ids)\n",
    "print(cnv_data_filtered)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "101b893a",
   "metadata": {},
   "source": [
    "Step 3: Create generic boolean networks with the sensitive and resistant ID names\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "509fe0ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_generic_models_cfg = 'models/generic_models/Adeno_lung_Cancer.cfg'\n",
    "folder_generic_models_bnd = 'models/generic_models/Adeno_lung_Cancer.bnd'\n",
    "# create_generic_patients_cfgs_bnds(folder_generic_models_cfg, folder_generic_models_bnd, folder_pers_models, top_resistant_ids, top_sensitive_ids, drug_interest, tissue_interest)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "446d42ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# specify the phenotypes of interest for the analysis\n",
    "\n",
    "phenotype_interest = [\"Proliferation\",\"Invasion\",\"DNA_Repair\",\"Migration\",\"Apoptosis\"]\n",
    "original_data_dir_res = \"models/personalized_boolean_Refametinib_PAN_CANCER/resistant_patient/generic_models\"\n",
    "results_dir_res = \"models/personalized_boolean_Refametinib_PAN_CANCER/resistant_patient/generic_models\"\n",
    "\n",
    "original_data_dir_sens = \"models/personalized_boolean_Refametinib_PAN_CANCER/sensitive_patient/generic_models\"\n",
    "results_dir_sens = \"models/personalized_boolean_Refametinib_PAN_CANCER/sensitive_patient/generic_models\"\n",
    "\n",
    "\n",
    "# generic_models_update_phenotypes(phenotype_interest, original_data_dir_res, results_dir_res)\n",
    "# generic_models_update_phenotypes(phenotype_interest, original_data_dir_sens, results_dir_sens)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca027a3d",
   "metadata": {},
   "source": [
    "Step 4: Personalize the cfg files with genes/ proteins\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "eb7c650c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "models/personalized_boolean_Refametinib_PAN_CANCER/resistant_patient/generic_models\n",
      "models/personalized_boolean_Refametinib_PAN_CANCER/resistant_patient/personalized_boolean_modified/models_gene_expression\n",
      "gene_expression_level                               High Gene Expression  \\\n",
      "model_id                                                                   \n",
      "SIDM00003              ATM, EEF2K, ERG, ETS1, ETV1, LDHA, PTCH1, RUNX...   \n",
      "SIDM00056              ATM, BAD, BAX, BMP2, CFLAR, EGFR, ETS1, ETV1, ...   \n",
      "SIDM00129                                          AXIN1, FADD, MYC, SMO   \n",
      "SIDM00175                           ATM, ATR, ETS1, MDM2, MXI1, NF1, SMO   \n",
      "SIDM00183              BIRC5, BRCA1, COX4I2, E2F1, EP300, ETV1, EZH2,...   \n",
      "\n",
      "gene_expression_level                                Low Gene Expression  \n",
      "model_id                                                                  \n",
      "SIDM00003                                                          FOXA1  \n",
      "SIDM00056                                        E2F1, EZH2, FGFR3, TERT  \n",
      "SIDM00129              APAF1, ATM, ATR, BIRC5, BRCA1, BRCA2, DAXX, EP...  \n",
      "SIDM00175                                                   APAF1, FOXA1  \n",
      "SIDM00183                                          CFLAR, ETS1, FOS, MYC  \n",
      "Refametinib\n"
     ]
    }
   ],
   "source": [
    "print(output_dir_resistant)\n",
    "print(output_dir_sec_resistant)\n",
    "\n",
    "print(table_rna_seq_patients.head())\n",
    "print(drug_interest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f6c25190",
   "metadata": {},
   "outputs": [],
   "source": [
    "# only three modif for resistant\n",
    "# personalized_patients_genes_cfgs(rna_seq_data, montagud_nodes, output_dir_resistant, output_dir_sec_resistant, patients_ids, table_rna_seq_patients, drug_interest)\n",
    "# personalized_patients_genes_cfgs(rna_seq_data, montagud_nodes, output_dir_sensitive, output_dir_sec_sensitive, patients_ids, table_rna_seq_patients, drug_interest)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bead9930",
   "metadata": {},
   "source": [
    "Step 5: Identification of which patients id have KRAS or EGFR mutation\n",
    "personalize the bnd files with the mutations common to Lung (KRAS/ EGFR):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "41d39de4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# personalized_patients_mutations_bnds(mutations_data,patients_ids,bnd_dir_res, drug_interest)\n",
    "# personalized_patients_mutations_bnds(mutations_data,patients_ids,bnd_dir_sens, drug_interest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "bdd9b96d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       model_id gene_symbol  total_copy_number cn_category\n",
      "2313  SIDM00003        BCL2                2.0        Loss\n",
      "2502  SIDM00003         SMO                5.0        Gain\n",
      "2570  SIDM00003         JUN                4.0        Gain\n",
      "2591  SIDM00003        EZH2                5.0        Gain\n",
      "2594  SIDM00003         ERG                4.0        Gain\n"
     ]
    }
   ],
   "source": [
    "print(cnv_data_filtered.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f624b2ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tailor_bnd_cnv_cm(cnv_data_filtered,bnd_dir_res, bnd_dir_sens, drug_interest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "28bb04fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tailor_bnd_tsg_onco_mut(mutations_data_filtered_combined,bnd_dir_res, bnd_dir_sens, drug_interest)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01754ed5",
   "metadata": {},
   "source": [
    "Step 6: compute the phenotype distribution "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "97592343",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- Results for patient: SIDM00719_Refametinib ---\n",
      "patients columns are: Index(['Apoptosis -- Invasion -- Proliferation', 'Apoptosis -- Proliferation',\n",
      "       'Apoptosis', 'Apoptosis -- Invasion', 'Apoptosis -- DNA_Repair',\n",
      "       'Apoptosis -- DNA_Repair -- Invasion -- Proliferation',\n",
      "       'Apoptosis -- DNA_Repair -- Proliferation'],\n",
      "      dtype='object')\n",
      "               Proliferation  Invasion  DNA_Repair  Migration  Apoptosis\n",
      "EGF_ON                   0.0       0.0         0.0        0.0        NaN\n",
      "FGF_ON                   0.0       0.0         0.0        0.0        NaN\n",
      "TGFb_ON                  0.0       0.0         0.0        0.0   0.000774\n",
      "Nutrients_ON             0.0       0.0         0.0        0.0        NaN\n",
      "Hypoxia_ON               0.0       0.0         0.0        0.0        NaN\n",
      "Acidosis_ON              0.0       0.0         0.0        0.0        NaN\n",
      "Androgen_ON              0.0       0.0         0.0        0.0        NaN\n",
      "TNFalpha_ON              0.0       0.0         0.0        0.0        NaN\n",
      "Carcinogen_ON            0.0       0.0         0.0        0.0        NaN\n",
      "\n",
      "--- Results for patient: SIDM01120_Refametinib ---\n",
      "patients columns are: Index(['Apoptosis -- Invasion', 'Invasion', 'Apoptosis',\n",
      "       'Apoptosis -- DNA_Repair -- Invasion', 'DNA_Repair -- Invasion'],\n",
      "      dtype='object')\n",
      "               Proliferation  Invasion  DNA_Repair  Migration  Apoptosis\n",
      "EGF_ON                   0.0    0.0072         0.0        0.0        NaN\n",
      "FGF_ON                   0.0    0.0072         0.0        0.0        NaN\n",
      "TGFb_ON                  0.0    0.0080         0.0        0.0        NaN\n",
      "Nutrients_ON             0.0    0.0086         0.0        0.0        NaN\n",
      "Hypoxia_ON               0.0    0.0096         0.0        0.0        NaN\n",
      "Acidosis_ON              0.0    0.0054         0.0        0.0     0.2578\n",
      "Androgen_ON              0.0    0.0072         0.0        0.0        NaN\n",
      "TNFalpha_ON              0.0    0.0072         0.0        0.0        NaN\n",
      "Carcinogen_ON            0.0       NaN         0.0        0.0        NaN\n",
      "\n",
      "--- Results for patient: SIDM00183_Refametinib ---\n",
      "patients columns are: Index(['Invasion', 'Invasion -- Proliferation', 'Apoptosis -- Invasion',\n",
      "       'Apoptosis -- Invasion -- Proliferation', 'DNA_Repair -- Invasion',\n",
      "       'DNA_Repair -- Invasion -- Proliferation'],\n",
      "      dtype='object')\n",
      "               Proliferation  Invasion  DNA_Repair  Migration  Apoptosis\n",
      "EGF_ON                   0.0  0.000157         0.0        0.0        0.0\n",
      "FGF_ON                   0.0  0.999400         0.0        0.0        0.0\n",
      "TGFb_ON                  0.0  1.000000         0.0        0.0        0.0\n",
      "Nutrients_ON             0.0  0.999400         0.0        0.0        0.0\n",
      "Hypoxia_ON               0.0  0.999800         0.0        0.0        0.0\n",
      "Acidosis_ON              0.0  0.999400         0.0        0.0        0.0\n",
      "Androgen_ON              0.0  0.000157         0.0        0.0        0.0\n",
      "TNFalpha_ON              0.0       NaN         0.0        0.0        0.0\n",
      "Carcinogen_ON            0.0       NaN         0.0        0.0        0.0\n",
      "\n",
      "--- Results for patient: SIDM00506_Refametinib ---\n",
      "patients columns are: Index(['<nil>', 'Apoptosis', 'Apoptosis -- Invasion -- Proliferation',\n",
      "       'Apoptosis -- Proliferation', 'Invasion', 'Invasion -- Proliferation',\n",
      "       'Proliferation', 'Apoptosis -- Invasion', 'Apoptosis -- DNA_Repair',\n",
      "       'Apoptosis -- DNA_Repair -- Invasion',\n",
      "       'Apoptosis -- DNA_Repair -- Invasion -- Proliferation',\n",
      "       'Apoptosis -- DNA_Repair -- Proliferation'],\n",
      "      dtype='object')\n",
      "               Proliferation  Invasion  DNA_Repair  Migration  Apoptosis\n",
      "EGF_ON              0.843526  0.006800         0.0        0.0   0.000400\n",
      "FGF_ON              0.312071  0.001800         0.0        0.0   0.061718\n",
      "TGFb_ON                  NaN       NaN         0.0        0.0   0.180266\n",
      "Nutrients_ON        0.834200  0.006433         0.0        0.0   0.000400\n",
      "Hypoxia_ON          0.837494  0.005200         0.0        0.0   0.000200\n",
      "Acidosis_ON         0.833299  0.004400         0.0        0.0   0.000200\n",
      "Androgen_ON         0.843326  0.006800         0.0        0.0   0.000400\n",
      "TNFalpha_ON              NaN  0.837398         0.0        0.0   0.000323\n",
      "Carcinogen_ON            NaN       NaN         0.0        0.0   0.075174\n",
      "\n",
      "--- Results for patient: SIDM00716_Refametinib ---\n",
      "patients columns are: Index(['Apoptosis -- Proliferation', 'Apoptosis -- Invasion',\n",
      "       'Apoptosis -- DNA_Repair -- Proliferation'],\n",
      "      dtype='object')\n",
      "               Proliferation  Invasion  DNA_Repair  Migration  Apoptosis\n",
      "EGF_ON                   0.0       0.0         0.0        0.0        0.0\n",
      "FGF_ON                   0.0       0.0         0.0        0.0        0.0\n",
      "TGFb_ON                  0.0       0.0         0.0        0.0        0.0\n",
      "Nutrients_ON             0.0       0.0         0.0        0.0        0.0\n",
      "Hypoxia_ON               0.0       0.0         0.0        0.0        0.0\n",
      "Acidosis_ON              0.0       0.0         0.0        0.0        0.0\n",
      "Androgen_ON              0.0       0.0         0.0        0.0        0.0\n",
      "TNFalpha_ON              0.0       0.0         0.0        0.0        0.0\n",
      "Carcinogen_ON            0.0       0.0         0.0        0.0        0.0\n",
      "\n",
      "--- Results for patient: SIDM00971_Refametinib ---\n",
      "patients columns are: Index(['Invasion -- Migration -- Proliferation', 'Invasion -- Proliferation',\n",
      "       'Proliferation', 'Apoptosis -- Invasion -- Migration -- Proliferation',\n",
      "       'Apoptosis -- Invasion -- Proliferation', 'Apoptosis -- Proliferation',\n",
      "       '<nil>', 'Apoptosis', 'Apoptosis -- Invasion', 'Invasion',\n",
      "       'Invasion -- Migration',\n",
      "       'Apoptosis -- DNA_Repair -- Invasion -- Migration -- Proliferation',\n",
      "       'Apoptosis -- DNA_Repair -- Invasion -- Proliferation',\n",
      "       'Apoptosis -- DNA_Repair -- Proliferation',\n",
      "       'DNA_Repair -- Invasion -- Migration -- Proliferation',\n",
      "       'DNA_Repair -- Invasion -- Proliferation',\n",
      "       'DNA_Repair -- Proliferation'],\n",
      "      dtype='object')\n",
      "               Proliferation  Invasion  DNA_Repair  Migration  Apoptosis\n",
      "EGF_ON              0.268800       NaN         0.0        0.0        NaN\n",
      "FGF_ON              0.245567       NaN         0.0        0.0        NaN\n",
      "TGFb_ON                  NaN  0.071844         0.0        0.0   0.206453\n",
      "Nutrients_ON        0.271600       NaN         0.0        0.0        NaN\n",
      "Hypoxia_ON               NaN  0.697907         0.0        0.0        NaN\n",
      "Acidosis_ON         0.265800       NaN         0.0        0.0        NaN\n",
      "Androgen_ON         0.269000       NaN         0.0        0.0        NaN\n",
      "TNFalpha_ON              NaN  0.763400         0.0        0.0        NaN\n",
      "Carcinogen_ON            NaN       NaN         0.0        0.0        NaN\n",
      "\n",
      "--- Results for patient: SIDM00881_Refametinib ---\n",
      "patients columns are: Index(['Proliferation', '<nil>', 'DNA_Repair -- Proliferation'], dtype='object')\n",
      "               Proliferation  Invasion  DNA_Repair  Migration  Apoptosis\n",
      "EGF_ON                 1.000       0.0         0.0        0.0        0.0\n",
      "FGF_ON                 1.000       0.0         0.0        0.0        0.0\n",
      "TGFb_ON                0.787       0.0         0.0        0.0        0.0\n",
      "Nutrients_ON           1.000       0.0         0.0        0.0        0.0\n",
      "Hypoxia_ON               NaN       0.0         0.0        0.0        0.0\n",
      "Acidosis_ON            1.000       0.0         0.0        0.0        0.0\n",
      "Androgen_ON            1.000       0.0         0.0        0.0        0.0\n",
      "TNFalpha_ON            1.000       0.0         0.0        0.0        0.0\n",
      "Carcinogen_ON            NaN       0.0         0.0        0.0        0.0\n",
      "\n",
      "--- Results for patient: SIDM00920_Refametinib ---\n",
      "patients columns are: Index(['<nil>', 'Invasion', 'Invasion -- Migration',\n",
      "       'Invasion -- Migration -- Proliferation', 'Invasion -- Proliferation',\n",
      "       'Proliferation', 'Apoptosis', 'Apoptosis -- Invasion',\n",
      "       'Apoptosis -- Invasion -- Migration',\n",
      "       'Apoptosis -- Invasion -- Migration -- Proliferation',\n",
      "       'Apoptosis -- Invasion -- Proliferation', 'Apoptosis -- Proliferation',\n",
      "       'Apoptosis -- DNA_Repair', 'Apoptosis -- DNA_Repair -- Invasion',\n",
      "       'Apoptosis -- DNA_Repair -- Invasion -- Proliferation',\n",
      "       'Apoptosis -- DNA_Repair -- Proliferation', 'DNA_Repair',\n",
      "       'DNA_Repair -- Invasion', 'DNA_Repair -- Invasion -- Migration',\n",
      "       'DNA_Repair -- Invasion -- Migration -- Proliferation',\n",
      "       'DNA_Repair -- Invasion -- Proliferation',\n",
      "       'DNA_Repair -- Proliferation'],\n",
      "      dtype='object')\n",
      "               Proliferation  Invasion  DNA_Repair  Migration  Apoptosis\n",
      "EGF_ON              0.059924  0.084322         NaN        0.0        NaN\n",
      "FGF_ON              0.045317  0.256272         NaN        0.0   0.013681\n",
      "TGFb_ON             0.002000  0.004454         NaN        0.0   0.017166\n",
      "Nutrients_ON        0.047180  0.318008         NaN        0.0        NaN\n",
      "Hypoxia_ON               NaN  0.541938         NaN        0.0        NaN\n",
      "Acidosis_ON         0.066729  0.284327         NaN        0.0        NaN\n",
      "Androgen_ON         0.059924  0.084322         NaN        0.0        NaN\n",
      "TNFalpha_ON         0.000873  0.588804         NaN        0.0        NaN\n",
      "Carcinogen_ON       0.000400  0.001950       0.001        0.0   0.110440\n",
      "\n",
      "--- Results for patient: SIDM00524_Refametinib ---\n",
      "patients columns are: Index(['Apoptosis -- Invasion -- Proliferation', 'Apoptosis -- Invasion',\n",
      "       'Apoptosis -- DNA_Repair -- Invasion',\n",
      "       'Apoptosis -- DNA_Repair -- Invasion -- Proliferation'],\n",
      "      dtype='object')\n",
      "               Proliferation  Invasion  DNA_Repair  Migration  Apoptosis\n",
      "EGF_ON                   0.0       0.0         0.0        0.0        0.0\n",
      "FGF_ON                   0.0       0.0         0.0        0.0        0.0\n",
      "TGFb_ON                  0.0       0.0         0.0        0.0        0.0\n",
      "Nutrients_ON             0.0       0.0         0.0        0.0        0.0\n",
      "Hypoxia_ON               0.0       0.0         0.0        0.0        0.0\n",
      "Acidosis_ON              0.0       0.0         0.0        0.0        0.0\n",
      "Androgen_ON              0.0       0.0         0.0        0.0        0.0\n",
      "TNFalpha_ON              0.0       0.0         0.0        0.0        0.0\n",
      "Carcinogen_ON            0.0       0.0         0.0        0.0        0.0\n",
      "\n",
      "--- Results for patient: SIDM00292_Refametinib ---\n",
      "patients columns are: Index(['Apoptosis', 'Apoptosis -- Proliferation', 'Apoptosis -- Invasion',\n",
      "       'Apoptosis -- Invasion -- Proliferation', 'Apoptosis -- DNA_Repair',\n",
      "       'Apoptosis -- DNA_Repair -- Proliferation'],\n",
      "      dtype='object')\n",
      "               Proliferation  Invasion  DNA_Repair  Migration  Apoptosis\n",
      "EGF_ON                   0.0       0.0         0.0        0.0     0.9994\n",
      "FGF_ON                   0.0       0.0         0.0        0.0     0.9994\n",
      "TGFb_ON                  0.0       0.0         0.0        0.0        NaN\n",
      "Nutrients_ON             0.0       0.0         0.0        0.0     0.9998\n",
      "Hypoxia_ON               0.0       0.0         0.0        0.0     0.9994\n",
      "Acidosis_ON              0.0       0.0         0.0        0.0     0.9994\n",
      "Androgen_ON              0.0       0.0         0.0        0.0     0.9994\n",
      "TNFalpha_ON              0.0       0.0         0.0        0.0        NaN\n",
      "Carcinogen_ON            0.0       0.0         0.0        0.0        NaN\n",
      "\n",
      "--- Results for patient: SIDM00870_Refametinib ---\n",
      "patients columns are: Index(['DNA_Repair -- Proliferation'], dtype='object')\n",
      "               Proliferation  Invasion  DNA_Repair  Migration  Apoptosis\n",
      "EGF_ON                   0.0       0.0         0.0        0.0        0.0\n",
      "FGF_ON                   0.0       0.0         0.0        0.0        0.0\n",
      "TGFb_ON                  0.0       0.0         0.0        0.0        0.0\n",
      "Nutrients_ON             0.0       0.0         0.0        0.0        0.0\n",
      "Hypoxia_ON               0.0       0.0         0.0        0.0        0.0\n",
      "Acidosis_ON              0.0       0.0         0.0        0.0        0.0\n",
      "Androgen_ON              0.0       0.0         0.0        0.0        0.0\n",
      "TNFalpha_ON              0.0       0.0         0.0        0.0        0.0\n",
      "Carcinogen_ON            0.0       0.0         0.0        0.0        0.0\n",
      "\n",
      "--- Results for patient: SIDM00508_Refametinib ---\n",
      "patients columns are: Index(['Apoptosis', 'Apoptosis -- Invasion', 'Apoptosis -- Proliferation',\n",
      "       'Apoptosis -- Invasion -- Proliferation', 'Apoptosis -- DNA_Repair',\n",
      "       'Apoptosis -- DNA_Repair -- Invasion',\n",
      "       'Apoptosis -- DNA_Repair -- Proliferation'],\n",
      "      dtype='object')\n",
      "               Proliferation  Invasion  DNA_Repair  Migration  Apoptosis\n",
      "EGF_ON                   0.0       0.0         0.0        0.0   0.986584\n",
      "FGF_ON                   0.0       0.0         0.0        0.0   0.987912\n",
      "TGFb_ON                  0.0       0.0         0.0        0.0   0.286579\n",
      "Nutrients_ON             0.0       0.0         0.0        0.0   0.988750\n",
      "Hypoxia_ON               0.0       0.0         0.0        0.0   0.990400\n",
      "Acidosis_ON              0.0       0.0         0.0        0.0   0.987912\n",
      "Androgen_ON              0.0       0.0         0.0        0.0   0.986584\n",
      "TNFalpha_ON              0.0       0.0         0.0        0.0   0.273176\n",
      "Carcinogen_ON            0.0       0.0         0.0        0.0        NaN\n",
      "\n",
      "--- Results for patient: SIDM00767_Refametinib ---\n",
      "patients columns are: Index(['Apoptosis -- Proliferation', 'Apoptosis -- Invasion',\n",
      "       'Apoptosis -- Invasion -- Proliferation',\n",
      "       'Apoptosis -- DNA_Repair -- Proliferation'],\n",
      "      dtype='object')\n",
      "               Proliferation  Invasion  DNA_Repair  Migration  Apoptosis\n",
      "EGF_ON                   0.0       0.0         0.0        0.0        0.0\n",
      "FGF_ON                   0.0       0.0         0.0        0.0        0.0\n",
      "TGFb_ON                  0.0       0.0         0.0        0.0        0.0\n",
      "Nutrients_ON             0.0       0.0         0.0        0.0        0.0\n",
      "Hypoxia_ON               0.0       0.0         0.0        0.0        0.0\n",
      "Acidosis_ON              0.0       0.0         0.0        0.0        0.0\n",
      "Androgen_ON              0.0       0.0         0.0        0.0        0.0\n",
      "TNFalpha_ON              0.0       0.0         0.0        0.0        0.0\n",
      "Carcinogen_ON            0.0       0.0         0.0        0.0        0.0\n",
      "\n",
      "--- Results for patient: SIDM01056_Refametinib ---\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyboardInterrupt\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[19]\u001b[39m\u001b[32m, line 5\u001b[39m\n\u001b[32m      3\u001b[39m phenotypes_list = phenotype_interest\n\u001b[32m      4\u001b[39m inputs_list = [\u001b[33m'\u001b[39m\u001b[33mEGF\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mFGF\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mTGFb\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mNutrients\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mHypoxia\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mAcidosis\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mAndrogen\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mTNFalpha\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mCarcinogen\u001b[39m\u001b[33m'\u001b[39m]\n\u001b[32m----> \u001b[39m\u001b[32m5\u001b[39m patient_res_data_dict = \u001b[43mcompute_phenotypes_distribution\u001b[49m\u001b[43m(\u001b[49m\u001b[43mphenotypes_list\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfolder_result\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdic_patient_resistant\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minputs_list\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdrug_interest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mresistant\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m      6\u001b[39m patient_sens_data_dict = compute_phenotypes_distribution(phenotypes_list, folder_result, dic_patient_sensitive, inputs_list, drug_interest, \u001b[33m'\u001b[39m\u001b[33msensitive\u001b[39m\u001b[33m'\u001b[39m)\n\u001b[32m      7\u001b[39m patients_res_df_mean, patients_res_df_std, stats_results_data_res_df = compute_mean_patients(patient_res_data_dict)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/repos/icr/Attractor-Resistance/MaBoSS_simulation/MaBoSS_phenotype_distribution.py:102\u001b[39m, in \u001b[36mcompute_phenotypes_distribution\u001b[39m\u001b[34m(phenotypes_list, folder, dir_files, inputs_list, drug_interest, group_label)\u001b[39m\n\u001b[32m    100\u001b[39m         \u001b[38;5;28;01mif\u001b[39;00m inactive_input != active_input:\n\u001b[32m    101\u001b[39m             model.network.set_istate(inactive_input, [\u001b[32m1\u001b[39m, \u001b[32m0\u001b[39m])\n\u001b[32m--> \u001b[39m\u001b[32m102\u001b[39m     result = \u001b[43mmodel\u001b[49m\u001b[43m.\u001b[49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    104\u001b[39m     last_state = result.get_last_states_probtraj()\n\u001b[32m    106\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/repos/icr/Attractor-Resistance/.env/lib/python3.13/site-packages/maboss/simulation.py:275\u001b[39m, in \u001b[36mSimulation.run\u001b[39m\u001b[34m(self, command, workdir, overwrite, prefix, cmaboss, only_final_state)\u001b[39m\n\u001b[32m    272\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m overwrite:\n\u001b[32m    273\u001b[39m     \u001b[38;5;28mself\u001b[39m.overwrite = overwrite\n\u001b[32m--> \u001b[39m\u001b[32m275\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mResult\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcommand\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mworkdir\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43moverwrite\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprefix\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/repos/icr/Attractor-Resistance/.env/lib/python3.13/site-packages/maboss/result.py:58\u001b[39m, in \u001b[36mResult.__init__\u001b[39m\u001b[34m(self, simul, command, workdir, overwrite, prefix)\u001b[39m\n\u001b[32m     55\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m command:\n\u001b[32m     56\u001b[39m     maboss_cmd = command\n\u001b[32m---> \u001b[39m\u001b[32m58\u001b[39m \u001b[38;5;28mself\u001b[39m._err = \u001b[43msubprocess\u001b[49m\u001b[43m.\u001b[49m\u001b[43mcall\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[43mmaboss_cmd\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43m-c\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_cfg\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43m-o\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m     59\u001b[39m \u001b[43m                            \u001b[49m\u001b[43mos\u001b[49m\u001b[43m.\u001b[49m\u001b[43mpath\u001b[49m\u001b[43m.\u001b[49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mprefix\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_bnd\u001b[49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m     60\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m._err:\n\u001b[32m     61\u001b[39m     \u001b[38;5;28mprint\u001b[39m(\u001b[33m\"\u001b[39m\u001b[33mError, MaBoSS returned non 0 value\u001b[39m\u001b[33m\"\u001b[39m, file=stderr)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/homebrew/Cellar/python@3.13/3.13.3/Frameworks/Python.framework/Versions/3.13/lib/python3.13/subprocess.py:397\u001b[39m, in \u001b[36mcall\u001b[39m\u001b[34m(timeout, *popenargs, **kwargs)\u001b[39m\n\u001b[32m    395\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m Popen(*popenargs, **kwargs) \u001b[38;5;28;01mas\u001b[39;00m p:\n\u001b[32m    396\u001b[39m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m--> \u001b[39m\u001b[32m397\u001b[39m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mp\u001b[49m\u001b[43m.\u001b[49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m=\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    398\u001b[39m     \u001b[38;5;28;01mexcept\u001b[39;00m:  \u001b[38;5;66;03m# Including KeyboardInterrupt, wait handled that.\u001b[39;00m\n\u001b[32m    399\u001b[39m         p.kill()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/homebrew/Cellar/python@3.13/3.13.3/Frameworks/Python.framework/Versions/3.13/lib/python3.13/subprocess.py:1277\u001b[39m, in \u001b[36mPopen.wait\u001b[39m\u001b[34m(self, timeout)\u001b[39m\n\u001b[32m   1275\u001b[39m     endtime = _time() + timeout\n\u001b[32m   1276\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m-> \u001b[39m\u001b[32m1277\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_wait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m=\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1278\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyboardInterrupt\u001b[39;00m:\n\u001b[32m   1279\u001b[39m     \u001b[38;5;66;03m# https://bugs.python.org/issue25942\u001b[39;00m\n\u001b[32m   1280\u001b[39m     \u001b[38;5;66;03m# The first keyboard interrupt waits briefly for the child to\u001b[39;00m\n\u001b[32m   1281\u001b[39m     \u001b[38;5;66;03m# exit under the common assumption that it also received the ^C\u001b[39;00m\n\u001b[32m   1282\u001b[39m     \u001b[38;5;66;03m# generated SIGINT and will exit rapidly.\u001b[39;00m\n\u001b[32m   1283\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/homebrew/Cellar/python@3.13/3.13.3/Frameworks/Python.framework/Versions/3.13/lib/python3.13/subprocess.py:2063\u001b[39m, in \u001b[36mPopen._wait\u001b[39m\u001b[34m(self, timeout)\u001b[39m\n\u001b[32m   2061\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.returncode \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m   2062\u001b[39m     \u001b[38;5;28;01mbreak\u001b[39;00m  \u001b[38;5;66;03m# Another thread waited.\u001b[39;00m\n\u001b[32m-> \u001b[39m\u001b[32m2063\u001b[39m (pid, sts) = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_try_wait\u001b[49m\u001b[43m(\u001b[49m\u001b[32;43m0\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m   2064\u001b[39m \u001b[38;5;66;03m# Check the pid and loop as waitpid has been known to\u001b[39;00m\n\u001b[32m   2065\u001b[39m \u001b[38;5;66;03m# return 0 even without WNOHANG in odd situations.\u001b[39;00m\n\u001b[32m   2066\u001b[39m \u001b[38;5;66;03m# http://bugs.python.org/issue14396.\u001b[39;00m\n\u001b[32m   2067\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m pid == \u001b[38;5;28mself\u001b[39m.pid:\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/homebrew/Cellar/python@3.13/3.13.3/Frameworks/Python.framework/Versions/3.13/lib/python3.13/subprocess.py:2021\u001b[39m, in \u001b[36mPopen._try_wait\u001b[39m\u001b[34m(self, wait_flags)\u001b[39m\n\u001b[32m   2019\u001b[39m \u001b[38;5;250m\u001b[39m\u001b[33;03m\"\"\"All callers to this function MUST hold self._waitpid_lock.\"\"\"\u001b[39;00m\n\u001b[32m   2020\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m-> \u001b[39m\u001b[32m2021\u001b[39m     (pid, sts) = \u001b[43mos\u001b[49m\u001b[43m.\u001b[49m\u001b[43mwaitpid\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mpid\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mwait_flags\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   2022\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mChildProcessError\u001b[39;00m:\n\u001b[32m   2023\u001b[39m     \u001b[38;5;66;03m# This happens if SIGCLD is set to be ignored or waiting\u001b[39;00m\n\u001b[32m   2024\u001b[39m     \u001b[38;5;66;03m# for child processes has otherwise been disabled for our\u001b[39;00m\n\u001b[32m   2025\u001b[39m     \u001b[38;5;66;03m# process.  This child is dead, we can't get the status.\u001b[39;00m\n\u001b[32m   2026\u001b[39m     pid = \u001b[38;5;28mself\u001b[39m.pid\n",
      "\u001b[31mKeyboardInterrupt\u001b[39m: "
     ]
    }
   ],
   "source": [
    "dic_patient_resistant =f'{folder_pers_models}/resistant_patient/personalized_boolean_modified/models_gene_expression'\n",
    "dic_patient_sensitive =f'{folder_pers_models}/sensitive_patient/personalized_boolean_modified/models_gene_expression'\n",
    "phenotypes_list = phenotype_interest\n",
    "inputs_list = ['EGF', 'FGF', 'TGFb', 'Nutrients', 'Hypoxia', 'Acidosis', 'Androgen', 'TNFalpha', 'Carcinogen']\n",
    "patient_res_data_dict = compute_phenotypes_distribution(phenotypes_list, folder_result, dic_patient_resistant, inputs_list, drug_interest, 'resistant')\n",
    "patient_sens_data_dict = compute_phenotypes_distribution(phenotypes_list, folder_result, dic_patient_sensitive, inputs_list, drug_interest, 'sensitive')\n",
    "patients_res_df_mean, patients_res_df_std, stats_results_data_res_df = compute_mean_patients(patient_res_data_dict)\n",
    "patients_sens_df_mean, patients_sens_df_std, stats_results_data_sens_df = compute_mean_patients(patient_sens_data_dict)\n",
    "patients_res_df_mean.to_csv(f'{folder_result}/resistant_results/only_gene_expression/single_input_on/patients_resistant_df_mean_{drug_interest}.csv', index=True)\n",
    "patients_res_df_std.to_csv(f'{folder_result}/resistant_results/only_gene_expression/single_input_on/patients_resistant_df_std_{drug_interest}.csv', index=True)\n",
    "stats_results_data_res_df.to_csv(f'{folder_result}/resistant_results/only_gene_expression/single_input_on/patients_resistant_values_stats_{drug_interest}.csv', index=True)\n",
    "stats_results_data_sens_df.to_csv(f'{folder_result}/sensitive_results/only_gene_expression/single_input_on/patients_sensitive_values_stats_{drug_interest}.csv', index=True)\n",
    "patients_sens_df_mean.to_csv(f'{folder_result}/sensitive_results/only_gene_expression/single_input_on/patients_sensitive_df_mean_{drug_interest}.csv', index=True)\n",
    "patients_sens_df_std.to_csv(f'{folder_result}/sensitive_results/only_gene_expression/single_input_on/patients_sensitive_df_std_{drug_interest}.csv', index=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "601cf228",
   "metadata": {},
   "source": [
    "Step 7: Compute stats test between two mean datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ceef3985",
   "metadata": {},
   "outputs": [],
   "source": [
    "patient_res_stats_values = pd.read_csv(f'{folder_result}/resistant_results/only_gene_expression/single_input_on/patients_resistant_values_stats_{drug_interest}.csv')\n",
    "patient_sens_stats_values = pd.read_csv(f'{folder_result}/sensitive_results/only_gene_expression/single_input_on/patients_sensitive_values_stats_{drug_interest}.csv')\n",
    "compute_mannwhitneyu_test_means(folder_result,patient_res_stats_values, patient_sens_stats_values, drug_interest)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf053c65",
   "metadata": {},
   "source": [
    "Step 8: Vizualise the boxplot of phenotype distribution output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "662bda1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "patient_res_values = pd.read_csv(f'{folder_result}/resistant_results/only_gene_expression/single_input_on/patients_resistant_values_stats_{drug_interest}.csv')\n",
    "patient_sens_values = pd.read_csv(f'{folder_result}/sensitive_results/only_gene_expression/single_input_on/patients_sensitive_values_stats_{drug_interest}.csv')\n",
    "data_greater_side = pd.read_csv(f'{folder_result}/sensitive_resistant_results/p_values_df_mannwhitneyu_greater_sign_{drug_interest}.csv')\n",
    "create_boxplot(folder_result, patient_res_values, patient_sens_values, data_greater_side)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "921ef689",
   "metadata": {},
   "source": [
    "Step 9: create table of patients with conditions- phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8259b991",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir_res_data = f'{folder_result}/resistant_results/only_gene_expression/single_input_on/phenotype_distribution_patients'\n",
    "dir_sens_data = f'{folder_result}/sensitive_results/only_gene_expression/single_input_on/phenotype_distribution_patients'\n",
    "patients_phenot_table = create_table_patients_phenotypes(folder_result, dir_res_data, dir_sens_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8320efc8",
   "metadata": {},
   "source": [
    "Step 10: Create heatmap figure "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5ebad17",
   "metadata": {},
   "outputs": [],
   "source": [
    "patient_resistant_mean = pd.read_csv(f'{folder_result}/resistant_results/only_gene_expression/single_input_on/patients_resistant_df_mean_{drug_interest}.csv')\n",
    "patient_sensitive_mean = pd.read_csv(f'{folder_result}/sensitive_results/only_gene_expression/single_input_on/patients_sensitive_df_mean_{drug_interest}.csv')\n",
    "vizualise_table_phenotype_condition(folder_result, patient_resistant_mean, patient_sensitive_mean)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e93b4134",
   "metadata": {},
   "source": [
    "Step 11: Identify genes differently expressed in the patients with high "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a417f999",
   "metadata": {},
   "outputs": [],
   "source": [
    "patients_phenot_table = pd.read_csv(f'{folder_result}/sensitive_resistant_results/patients_phenot_table.csv')\n",
    "genes_stats_results_metast_TGFb,  len_resistant_group_id_metast_TGFb, len_sensitive_group_ids_metast_TGFb = compute_genes_mean_signature(rna_seq_data_filtered, folder_result, 'Metastasis', 'TGFb', patients_phenot_table, top_resistant_ids, top_sensitive_ids)\n",
    "genes_stats_results_prolif_egf,  len_resistant_group_ids_prolif_egf, len_sensitive_group_ids_prolif_egf = compute_genes_mean_signature(rna_seq_data_filtered, folder_result, 'Proliferation', 'EGF', patients_phenot_table, top_resistant_ids, top_sensitive_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d139e43",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len_resistant_group_id_metast_TGFb)\n",
    "print(len_sensitive_group_ids_metast_TGFb)\n",
    "\n",
    "print(len_resistant_group_ids_prolif_egf)\n",
    "print(len_sensitive_group_ids_prolif_egf)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "478874e0",
   "metadata": {},
   "source": [
    "Step 11: check there is not correlation between phenotype distribution and cancer type \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af833d63",
   "metadata": {},
   "outputs": [],
   "source": [
    "# patients_phenot_table['SANGER_MODEL_ID'] = patients_phenot_table['Unnamed: 0'].str.split('_').str[0]\n",
    "# conditions = [\n",
    "#     patients_phenot_table['SANGER_MODEL_ID'].isin(top_resistant_ids),\n",
    "#     patients_phenot_table['SANGER_MODEL_ID'].isin(top_sensitive_ids)\n",
    "#     ]\n",
    "# choices = ['Resistant', 'Sensitive']\n",
    "# patients_phenot_table.loc[:,'Drug status'] = np.select(conditions, choices, default = '')\n",
    "\n",
    "\n",
    "\n",
    "# ids_tissue_data = drug_tissue_data[['SANGER_MODEL_ID', 'tissue']]\n",
    "# ids_tissue_data = ids_tissue_data.drop_duplicates(subset='SANGER_MODEL_ID')\n",
    "\n",
    "\n",
    "# # merge tissues and model id \n",
    "# patients_phenot_table = pd.merge(patients_phenot_table, ids_tissue_data, on = 'SANGER_MODEL_ID')\n",
    "\n",
    "\n",
    "\n",
    "# # look the number of each cancer for the condition-phenotype of interest\n",
    "# condition = 'TGFb'\n",
    "# phenotype = 'Metastasis'\n",
    "\n",
    "\n",
    "# # resistant group changes according to what is the condition and the phenotype\n",
    "# # group_proliferation_resistant: group with high phenotype \n",
    "\n",
    "# group_phenotype_resistant = patients_phenot_table[\n",
    "#     (patients_phenot_table['Drug status'] == 'Resistant') & \n",
    "#     (patients_phenot_table[f'{condition}_ON_{phenotype}'] >= 0.1)\n",
    "# ]\n",
    "\n",
    "# # print(group_phenotype_resistant['tissue'].value_counts()) # EGF- proliferation: 4 lung, 1 breast, 1 haematopoetic\n",
    "# #                                                           # TGFb- Metastasis: 21 haemato, 2 skin, 2 breast, 1 lung, 1 large intestine, 1 endom, 1 liver\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a51e26a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# inputs = ['EGF', 'FGF', 'Nutrients', 'Androgen']\n",
    "# results = []\n",
    "# output_dir = f'{folder_result}/sensitive_results/only_gene_expression/all_inputs_on/phenotype_distribution_patients'\n",
    "# os.makedirs(output_dir, exist_ok=True)\n",
    "# cfg_files = [f for f in os.listdir(dic_patient_sensitive) if f.endswith(f\"{drug_interest}.cfg\")]\n",
    "# for cfg_file in cfg_files:\n",
    "#     cfg_path = os.path.join(dic_patient_sensitive, cfg_file)\n",
    "#     base_name = os.path.splitext(cfg_file)[0]\n",
    "#     bnd_path = os.path.join(dic_patient_sensitive, base_name + \".bnd\")\n",
    "#     model = maboss.load(bnd_path, cfg_path)\n",
    "#     for input_node in inputs:\n",
    "#         model.network.set_istate(input_node, [0.0, 1.0])  # ON\n",
    "#     result = model.run()\n",
    "#     probtraj = result.get_last_states_probtraj()\n",
    "#     probtraj[\"Patient_ID\"] = base_name\n",
    "#     results.append(probtraj)\n",
    "# combined_df = pd.concat(results, ignore_index=True)\n",
    "# combined_csv_path = os.path.join(output_dir, \"all_patients_all_inputs_on_probtraj.csv\")\n",
    "# combined_df.to_csv(combined_csv_path, index=False)\n",
    "# print(f\"Saved combined results to: {combined_csv_path}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb8ce568",
   "metadata": {},
   "outputs": [],
   "source": [
    "# results_resistant = pd.read_csv('results/resistant_results/only_gene_expression/all_inputs_on/phenotype_distribution_patients/all_patients_all_inputs_on_probtraj.csv')\n",
    "# results_sensitive = pd.read_csv('results/sensitive_results/only_gene_expression/all_inputs_on/phenotype_distribution_patients/all_patients_all_inputs_on_probtraj.csv')\n",
    "\n",
    "\n",
    "\n",
    "# print('Apoptosis')\n",
    "\n",
    "# resistant_apopto_mean = results_resistant['Apoptosis'].mean()\n",
    "# print(resistant_apopto_mean)\n",
    "\n",
    "# sensitive_apopto_mean = results_sensitive['Apoptosis'].mean()\n",
    "# print(sensitive_apopto_mean)\n",
    "\n",
    "\n",
    "\n",
    "# print('Proliferation')\n",
    "# resistant_prolif_mean = results_resistant['Proliferation'].mean()\n",
    "# print(resistant_prolif_mean)\n",
    "\n",
    "# sensitive_prolif_mean = results_sensitive['Proliferation'].mean()\n",
    "# print(sensitive_prolif_mean)\n",
    "\n",
    "\n",
    "\n",
    "# print('Metastasis')\n",
    "# resistant_metas_mean = results_resistant['Metastasis'].mean()\n",
    "# print(resistant_metas_mean)\n",
    "\n",
    "# sensitive_metas_mean = results_sensitive['Metastasis'].mean()\n",
    "# print(sensitive_metas_mean)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f388be28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TO DO: later- changes EGF and TGF initial states according to expression\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
