{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6658aef1",
   "metadata": {},
   "source": [
    "# Pipeline to create the personalized Boolean Networks based on gene expression and Copy Number Variation Data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "edb31b1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script type=\"text/javascript\" id=\"colomoto-setup-4322361\" class=\"to-be-removed\">\n",
       "        if (typeof Jupyter != 'undefined') {\n",
       "            \n",
       "function detect_import(cell, module) {\n",
       "    var code = cell.get_text();\n",
       "    code = code.replace(/\\\\\\n/g, \"\");\n",
       "    var lines = code.split(\"\\n\");\n",
       "    var r_simple = new RegExp(\"^(\"+module+\")$\");\n",
       "    var r_alias = new RegExp(\"^\"+module+\"\\\\s+as\\\\s+(\\\\w+)$\");\n",
       "    for (var i = 0; i < lines.length; ++i) {\n",
       "        if (/^import\\s/.test(lines[i])) {\n",
       "            code = lines[i].substr(7);\n",
       "            var parts = code.split(\",\")\n",
       "            for (var j = 0; j < parts.length; ++j) {\n",
       "                code = parts[j].trim();\n",
       "                var m = code.match(r_simple);\n",
       "                if (!m) {\n",
       "                    m = code.match(r_alias);\n",
       "                }\n",
       "                if (m) {\n",
       "                    return m[1];\n",
       "                }\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "    return module;\n",
       "}\n",
       "\n",
       "function colomoto_replace_call(cell, orig, dest, args, comment=false) {\n",
       "    var call_regexp = new RegExp(\"\\\\.\"+orig.replace(\".\",\"\\\\.\")+\"\\\\(\");\n",
       "    var call_replacer = new RegExp(\"\\\\.\"+orig.replace(\".\",\"\\\\.\")\n",
       "                    + \"\\\\(\\\\s*([^\\\\)]*)?\\\\)\");\n",
       "    var code = cell.get_text();\n",
       "    var lines = code.split(\"\\n\");\n",
       "    if (args) {\n",
       "        var strargs = \", \"+args.join(\", \");\n",
       "    } else {\n",
       "        var strargs = \"\";\n",
       "    }\n",
       "    for (var i = 0; i < lines.length; ++i) {\n",
       "        if (call_regexp.test(lines[i])) {\n",
       "            var code = \"\"\n",
       "            if (comment) {\n",
       "                code += \"#\"+lines[i]+\"\\n\";\n",
       "            }\n",
       "            code += lines[i].replace(call_replacer, \".\"+dest+\"($1\"+strargs+\")\")\n",
       "            lines[i] = code;\n",
       "        }\n",
       "    }\n",
       "    cell.set_text(code)\n",
       "}\n",
       "\n",
       "function colomoto_upload(Jupyter, ssid, input, py_callback_name, orig, dest) {\n",
       "\n",
       "    function callback(out_data) {\n",
       "        var cell_element = $(\"#\"+ssid).parents('.cell');\n",
       "        var cell_idx = Jupyter.notebook.get_cell_elements().index(cell_element);\n",
       "        var cell = Jupyter.notebook.get_cell(cell_idx);\n",
       "\n",
       "        var filename = out_data.content.text;\n",
       "\n",
       "        var code = cell.get_text();\n",
       "        code = code.replace(new RegExp(\"\\\\b\" + orig.replace('.', '\\\\.')\n",
       "                    + \"\\\\(\\\\s*((\\\\w+)=[^\\\\)]*)?\\\\)\"),\n",
       "                dest+\"(\\\"\"+filename+\"\\\",$1)\");\n",
       "        code = code.replace('\",)', '\")')\n",
       "        cell.set_text(code);\n",
       "\n",
       "        Jupyter.notebook.select(cell_idx);\n",
       "        Jupyter.notebook.execute_cell_and_select_below();\n",
       "    }\n",
       "\n",
       "    if (! (window.File && window.FileReader && window.FileList && window.Blob)) {\n",
       "        alert(\"Interactive file upload is not supported by your browser.\");\n",
       "        return;\n",
       "    }\n",
       "\n",
       "    input.disabled = true;\n",
       "    input.style.cursor = \"wait\";\n",
       "    input.parentElement.style.cursor = \"wait\";\n",
       "\n",
       "    var f = input.files[0];\n",
       "    var reader = new FileReader();\n",
       "    reader.onload = (function(f) {\n",
       "        return function (e) {\n",
       "            var obj = {\n",
       "                content: e.target.result,\n",
       "                name: f.name\n",
       "            };\n",
       "\n",
       "            //var pycb = py_callback_name+\"(\"+JSON.stringify(obj)+\")\"\n",
       "            // hack/workaround:\n",
       "            // it seems that Jupyter does not like very long lines\n",
       "            // so we split the data in chunks\n",
       "            var chunk_length = 100;\n",
       "            var pycb = \"__colomoto_upload_name = \" + JSON.stringify(obj.name) + \"\\n\";\n",
       "            pycb += \"__colomoto_upload_content = \\\\\\n\";\n",
       "            for (var i = 0; i < obj.content.length; i += chunk_length) {\n",
       "                pycb += \"\\\"\" + obj.content.substr(i, chunk_length)+\"\\\"\\\\\\n\"\n",
       "            }\n",
       "            pycb += \"\\n\"\n",
       "            pycb += py_callback_name+\"({'name':__colomoto_upload_name, 'content': __colomoto_upload_content})\\n\";\n",
       "            pycb += \"del __colomoto_upload_name, __colomoto_upload_content\"\n",
       "\n",
       "            IPython.notebook.kernel.execute(pycb, {iopub: {output: callback}});\n",
       "        };\n",
       "    })(f);\n",
       "    reader.readAsDataURL(f);\n",
       "}\n",
       "\n",
       "function resolve_function(tool_api, funcname) {\n",
       "    if (tool_api.hasOwnProperty(funcname)) {\n",
       "        return tool_api[funcname];\n",
       "    } else {\n",
       "        return window[funcname];\n",
       "    }\n",
       "}\n",
       "\n",
       "function colomoto_extension(Jupyter, ssid, name, menu, toolbar, tool_api) {\n",
       "\n",
       "    function insert_snippet_code(snippet) {\n",
       "        var cell = Jupyter.notebook.get_selected_cell();\n",
       "        Jupyter.notebook.edit_mode();\n",
       "        cell.code_mirror.replaceSelection(snippet, 'around');\n",
       "        //cell.focus_editor();\n",
       "    }\n",
       "\n",
       "    /**\n",
       "        from https://github.com/moble/jupyter_boilerplate/blob/master/main.js\n",
       "    */\n",
       "    function callback_insert_snippet (evt) {\n",
       "        // this (or event.currentTarget, see below) always refers to the DOM\n",
       "        // element the listener was attached to - see\n",
       "        // http://stackoverflow.com/questions/12077859\n",
       "        insert_snippet_code($(evt.currentTarget).data('snippet-code'));\n",
       "    }\n",
       "    function build_menu_element (menu_item_spec, direction) {\n",
       "        // Create the menu item html element\n",
       "        var element = $('<li/>');\n",
       "\n",
       "        if (typeof menu_item_spec == 'string') {\n",
       "            if (menu_item_spec != '---') {\n",
       "                return element.html(menu_item_spec)\n",
       "                        .addClass('ui-state-disabled')\n",
       "                        .attr({\"style\": \"padding:2px .4em\"})\n",
       "                       ;\n",
       "            }\n",
       "            return element.addClass('divider');\n",
       "        }\n",
       "\n",
       "        var a = $('<a/>')\n",
       "            .attr('href', '#')\n",
       "            .html(menu_item_spec.name)\n",
       "            .appendTo(element);\n",
       "        if (menu_item_spec.hasOwnProperty('snippet')) {\n",
       "            var snippet = menu_item_spec.snippet;\n",
       "            if (typeof snippet == 'string' || snippet instanceof String) {\n",
       "                snippet = [snippet];\n",
       "            }\n",
       "            a.attr({\n",
       "                'title' : \"\", // Do not remove this, even though it's empty!\n",
       "                'data-snippet-code' : snippet.join('\\n'),\n",
       "            })\n",
       "            .on('click', callback_insert_snippet)\n",
       "            .addClass('snippet');\n",
       "        }\n",
       "        else if (menu_item_spec.hasOwnProperty('internal-link')) {\n",
       "            a.attr('href', menu_item_spec['internal-link']);\n",
       "        }\n",
       "        else if (menu_item_spec.hasOwnProperty('external-link')) {\n",
       "            a.empty();\n",
       "            a.attr('href', menu_item_spec['external-link']);\n",
       "            a.attr({\n",
       "                'target' : '_blank',\n",
       "                'title' : 'Opens in a new window',\n",
       "            });\n",
       "            $('<i class=\"fa fa-external-link menu-icon pull-right\"/>').appendTo(a);\n",
       "            $('<span/>').html(menu_item_spec.name).appendTo(a);\n",
       "        }\n",
       "\n",
       "        if (menu_item_spec.hasOwnProperty('sub-menu')) {\n",
       "            element\n",
       "                .addClass('dropdown-submenu')\n",
       "                .toggleClass('dropdown-submenu-left', direction === 'left');\n",
       "            var sub_element = $('<ul class=\"dropdown-menu\"/>')\n",
       "                .toggleClass('dropdown-menu-compact', menu_item_spec.overlay === true) // For space-saving menus\n",
       "                .appendTo(element);\n",
       "\n",
       "            var new_direction = (menu_item_spec['sub-menu-direction'] === 'left') ? 'left' : 'right';\n",
       "            for (var j=0; j<menu_item_spec['sub-menu'].length; ++j) {\n",
       "                var sub_menu_item_spec = build_menu_element(menu_item_spec['sub-menu'][j], new_direction);\n",
       "                if(sub_menu_item_spec !== null) {\n",
       "                    sub_menu_item_spec.appendTo(sub_element);\n",
       "                }\n",
       "            }\n",
       "        }\n",
       "\n",
       "        return element;\n",
       "    }\n",
       "\n",
       "    function menu_setup (menu_item_specs, sibling, insert_before_sibling) {\n",
       "        for (var i=0; i<menu_item_specs.length; ++i) {\n",
       "            var menu_item_spec;\n",
       "            if (insert_before_sibling) {\n",
       "                menu_item_spec = menu_item_specs[i];\n",
       "            } else {\n",
       "                menu_item_spec = menu_item_specs[menu_item_specs.length-1-i];\n",
       "            }\n",
       "            var direction = (menu_item_spec['menu-direction'] == 'left') ? 'left' : 'right';\n",
       "            var menu_element = build_menu_element(menu_item_spec, direction);\n",
       "            // We need special properties if this item is in the navbar\n",
       "            if ($(sibling).parent().is('ul.nav.navbar-nav')) {\n",
       "                menu_element\n",
       "                    .addClass('dropdown')\n",
       "                    .removeClass('dropdown-submenu dropdown-submenu-left');\n",
       "                menu_element.children('a')\n",
       "                    .addClass('dropdown-toggle')\n",
       "                    .attr({\n",
       "                        'id': name+'_menu',\n",
       "                        'data-toggle' : 'dropdown',\n",
       "                        'aria-expanded' : 'false'\n",
       "                    });\n",
       "            }\n",
       "\n",
       "            // Insert the menu element into DOM\n",
       "            menu_element[insert_before_sibling ? 'insertBefore': 'insertAfter'](sibling);\n",
       "        }\n",
       "    }\n",
       "    /** end from */\n",
       "\n",
       "\n",
       "    function self_cleanup() {\n",
       "        var cell_element = $(\"script[class='to-be-removed']\").parents('.cell');\n",
       "        var cell_idx = Jupyter.notebook.get_cell_elements().index(cell_element);\n",
       "        var cell = Jupyter.notebook.get_cell(cell_idx);\n",
       "        var to_remove = -1;\n",
       "        for (var i = 0; i < cell.output_area.outputs.length; ++i) {\n",
       "            var oa = cell.output_area.outputs[i];\n",
       "            if (oa.output_type == \"display_data\"\n",
       "                && typeof oa.data[\"text/html\"] != 'undefined'\n",
       "                && oa.data[\"text/html\"].indexOf(' class=\"to-be-removed\"') >= 0) {\n",
       "                to_remove = i;\n",
       "                break;\n",
       "            }\n",
       "        }\n",
       "        if (to_remove == -1) {\n",
       "            console.log(\"cannot find toberemoved\");\n",
       "        } else {\n",
       "            cell.output_area.outputs.splice(to_remove, 1);\n",
       "        }\n",
       "    }\n",
       "\n",
       "    function toolbar_setup(actions) {\n",
       "        var buttons = [];\n",
       "        for (var i = 0; i < actions.length; ++i) {\n",
       "            var setup = actions[i].setup;\n",
       "            if (typeof setup.handler == 'string') {\n",
       "                setup.handler = resolve_function(tool_api, setup.handler);\n",
       "            }\n",
       "            buttons.push(Jupyter.actions.register(actions[i].setup,\n",
       "                actions[i].name, name));\n",
       "        }\n",
       "        $(\"#\"+name+\"-toolbar\").remove();\n",
       "        Jupyter.toolbar.add_buttons_group(buttons, name+\"-toolbar\");\n",
       "    }\n",
       "\n",
       "    function replace_menu_snippets(menu_spec, orig, dest) {\n",
       "        if (menu_spec.hasOwnProperty(\"snippet\")) {\n",
       "            var snippet = menu_spec.snippet;\n",
       "            if (typeof snippet == \"string\" || snippet instanceof String) {\n",
       "                menu_spec[\"snippet\"] = snippet.replace(orig, dest);\n",
       "            } else {\n",
       "                for (var i = 0; i < snippet.length; ++i) {\n",
       "                    menu_spec[\"snippet\"][i] = snippet[i].replace(orig, dest);\n",
       "                }\n",
       "            }\n",
       "        }\n",
       "        if (menu_spec.hasOwnProperty(\"sub-menu\")) {\n",
       "            for (var i = 0; i < menu_spec[\"sub-menu\"].length; ++i) {\n",
       "                replace_menu_snippets(menu_spec[\"sub-menu\"][i], orig, dest);\n",
       "            }\n",
       "        }\n",
       "    }\n",
       "\n",
       "    function load_ipython_extension() {\n",
       "\n",
       "        var mycellelt = $(\"#\"+ssid).parents('.cell');\n",
       "        var myidx = Jupyter.notebook.get_cell_elements().index(mycellelt);\n",
       "        var import_cell = Jupyter.notebook.get_cell(myidx);\n",
       "\n",
       "        var alias = detect_import(import_cell, name);\n",
       "        tool_api.module_alias = alias;\n",
       "        if (alias && alias != name) {\n",
       "            var orig = new RegExp(\"\\\\b\"+name+\"\\\\b\", \"g\");\n",
       "            replace_menu_snippets(menu, orig, alias);\n",
       "        }\n",
       "\n",
       "        if (toolbar) {\n",
       "            toolbar_setup(toolbar);\n",
       "        }\n",
       "\n",
       "        $(\"#\"+name+\"_menu\").parent().remove();\n",
       "        if (menu) {\n",
       "            menu_setup([menu], $(\"#help_menu\").parent(), true);\n",
       "        }\n",
       "\n",
       "        if (tool_api.hasOwnProperty(\"post_install_callback\")) {\n",
       "            tool_api.post_install_callback();\n",
       "        }\n",
       "\n",
       "        setTimeout(self_cleanup, 5000);\n",
       "    };\n",
       "\n",
       "    load_ipython_extension();\n",
       "}\n",
       "\n",
       "function resolve_toolbar_handlers(tool_api, toolbar_spec) {\n",
       "    for (var i = 0; i < toolbar_spec.length; ++i) {\n",
       "        func = resolve_function(tool_api, toolbar_spec[i][\"setup\"][\"handler\"]);\n",
       "        toolbar_spec[i][\"setup\"][\"handler\"] = func;\n",
       "    }\n",
       "    return toolbar_spec\n",
       "}\n",
       "\n",
       "\n",
       "    var minibn_jsapi = {  };\n",
       "    colomoto_extension(Jupyter, \"colomoto-setup-4322361\", \"minibn\", null, null, minibn_jsapi);\n",
       "     }</script>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<script type=\"text/javascript\" id=\"colomoto-setup-3486999\" class=\"to-be-removed\">\n",
       "        if (typeof Jupyter != 'undefined') {\n",
       "            \n",
       "    var mymodule_jsapi = {  };\n",
       "    colomoto_extension(Jupyter, \"colomoto-setup-3486999\", \"mymodule\", {\"name\": \"MaBoSS\", \"sub-menu\": [{\"name\": \"Load MaBoSS file\", \"snippet\": [\"masim = maboss.load(\\\"filename.bnd\\\", \\\"filename.cfg\\\")\"]}, {\"name\": \"Network\", \"sub-menu\": [{\"name\": \"Set istate\", \"snippet\": [\"maboss.wg_set_istate(masim)\"]}, {\"name\": \"Get initial state\", \"snippet\": [\"masim.get_initial_state()\"]}]}, {\"name\": \"Simulation\", \"sub-menu\": [{\"name\": \"Create mutant\", \"snippet\": [\"maboss.wg_make_mutant(masim)\"]}, {\"name\": \"Run\", \"snippet\": [\"simres = masim.run()\"]}, {\"name\": \"Set initial states\", \"snippet\": [\"masim.network.set_istate([\\\"mygenelist\\\"],{})\"]}, {\"name\": \"Set output\", \"snippet\": [\"maboss.wg_set_output(masim)\"]}]}, {\"name\": \"Results\", \"sub-menu\": [{\"name\": \"Save results\", \"snippet\": [\"simres.save(\\\"filename\\\")\"]}, {\"name\": \"Plot piechart\", \"snippet\": [\"simres.plot_piechart()\"]}, {\"name\": \"Plot trajectory\", \"snippet\": [\"simres.plot_trajectory()\"]}]}, {\"name\": \"Conversion\", \"sub-menu\": [{\"name\": \"Convert to biolqm\", \"snippet\": [\"biolqm_model = maboss.to_biolqm(masim)\"]}, {\"name\": \"Convert to minibn\", \"snippet\": [\"minibn_model = maboss.to_minibn(masim)\"]}]}, \"---\", {\"name\": \"Documentation\", \"external-link\": \"http://pymaboss.readthedocs.io\"}]}, null, mymodule_jsapi);\n",
       "     }</script><style type=\"text/css\">\n",
       "    #mymodule_menu {\n",
       "        color: green;\n",
       "        background-color: white;\n",
       "    }\n",
       "    </style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import shutil\n",
    "\n",
    "from functions.generate_utils.identification_patients.get_patients_sens_res import get_patients\n",
    "\n",
    "from functions.generate_utils.identification_drug.drug_analysis import identify_drug\n",
    "\n",
    "# from functions.generate_utils.create_generic_models.update_nodes_names import replace_node_names_in_file\n",
    "\n",
    "\n",
    "from functions.generate_utils.create_generic_models.update_nodes_names import replace_node_names_in_file\n",
    "\n",
    "\n",
    "from functions.analysis_utils.MaBoSS_simulation.maboss_phenotype_patient import compute_phenotype_table, collect_group_data\n",
    "\n",
    "from functions.analysis_utils.stats.stats_proba import compute_mannwhitneyu_test_means\n",
    "from functions.analysis_utils.results_MaBoSS_visualization.boxplot_phenotype import create_boxplot\n",
    "\n",
    "\n",
    "from functions.generate_models import generate_models_re, pre_process_re\n",
    "\n",
    "from functions.analysis import downstream_analysis\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70927b0e",
   "metadata": {},
   "source": [
    "### Loading the datasets "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "69949123",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/83/pqwsm5cn7vscbq8l3fnf30qr0000gn/T/ipykernel_78660/496007439.py:17: DtypeWarning: Columns (22,50,56,57,58,59,61) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  mutations_data_depseq = pd.read_csv('data/depmap_data/OmicsSomaticMutations.csv')\n"
     ]
    }
   ],
   "source": [
    "# Loading the data (TCGA, cell model passport)\n",
    "\n",
    "annotations_models = pd.read_csv('data/model_list_20250407.csv')\n",
    "mutations_data = pd.read_csv('data/mutations_all_20250318.csv')\n",
    "drug_data = pd.read_csv('data/drug_sensitivity.csv')\n",
    "\n",
    "montagud_data = (\n",
    "    pd.read_csv('data/Montagud_inter_nodes_data.csv', header=1)\n",
    "    .loc[:, ['Target node', 'Interaction type', 'Source']])\n",
    "rna_seq_data = pd.read_csv('data/rnaseq_merged/rnaseq_merged_20250117.csv')\n",
    "cnv_data = pd.read_csv('data/cellmodel_data/cnv_summary_20250207.csv')\n",
    "\n",
    "proteins_data = pd.read_csv('data/cellmodel_data/proteomics_all_20250211.csv')\n",
    "\n",
    "\n",
    "models_depmap_annotation = pd.read_csv('data/cellmodel_data/model_list_20250423.csv')\n",
    "mutations_data_depseq = pd.read_csv('data/depmap_data/OmicsSomaticMutations.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e9802f53",
   "metadata": {},
   "outputs": [],
   "source": [
    "# identify_drug(drug_data, annotations_models, tissue_remove)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ce13899",
   "metadata": {},
   "source": [
    "### Users parameters data "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f811a1dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# User's data to choose\n",
    "\n",
    "# example\n",
    "type_model = 'genes_proteins_models'\n",
    "drug_interest = 'Refametinib'\n",
    "drug_targets = ['MEK1', 'MEK2']\n",
    "\n",
    "\n",
    "\n",
    "number_patients = 35 # in each group\n",
    "\n",
    "\n",
    "tissue_interest = 'PAN_CANCER'\n",
    "\n",
    "\n",
    "\n",
    "# Generic Boolean Network \n",
    "# Inputs\n",
    "inputs_list = ['EGF', 'FGF', 'TGFB', 'ANDROGEN', 'HYPOXIA', 'NUTRIENTS', 'CARCINOGEN', 'ACIDOSIS', 'TNF', 'SPOP']\n",
    "\n",
    "# Phenotypes\n",
    "phenotype_interest = [\"PROLIFERATION\",\"INVASION\",\"DNA_REPAIR\",\"APOPTOSIS\"]\n",
    "\n",
    "\n",
    "# to specify the boolean networks to be genes or proteins based\n",
    "name_montagud_maps = {\n",
    "        \"CASPASE8\": \"CASP8\",\n",
    "        \"CASPASE3\": \"CASP3\",\n",
    "        \"CASPASE9\": \"CASP9\",\n",
    "        \"CYCLINB\": \"CCNB1\",\n",
    "        \"CYCLIND\": \"CCND1\",\n",
    "        \"DSH\": \"DVL1\",\n",
    "        \"BETA_CATENIN\": \"CTNNB1\",\n",
    "        \"E_CADHERIN\": \"CDH1\",\n",
    "        \"CYCC\": \"CYCS\",\n",
    "        \"MEK1_2\": \"MEK1\",\n",
    "        \"NF_KB\": \"NFKB\",\n",
    "        \"SNAIL\": \"SNAI1\",\n",
    "        \"TNFALPHA\": \"TNF\",\n",
    "        \"TSC1_2\": \"TSC1\",\n",
    "        \"BCL_XL\": \"BCL2L1\",\n",
    "        \"MAP3K1_3\": \"MAP3K1\",\n",
    "        \"CHK1_2\": \"CHK1\",\n",
    "    }\n",
    "\n",
    "nodes_to_add = ['MEK2', 'TSC2', 'MAP3K3', 'CHK2']\n",
    "node_to_remove = ['FUSED_EVENT', 'AR_ERG']\n",
    "\n",
    "\n",
    "\n",
    "# proteins synonyms \n",
    "synonyms_maps = {\n",
    "        \"RAF1\": \"RAF\",\n",
    "        \"BCL2L2\": \"BCL2\",\n",
    "        \"SMAD1\": \"SMAD\",\n",
    "        \"SMAD3\": \"SMAD\",\n",
    "        \"SMAD4\": \"SMAD\",\n",
    "        \"SMAD5\": \"SMAD\",\n",
    "        \"BAK1\": \"BAK\",\n",
    "        \"GSK3A\": \"GSK3\",\n",
    "        \"GSK3B\": \"GSK3\",\n",
    "        \"TGFBR2\": \"TGFBR\",\n",
    "        \"TGFBR3\": \"TGFBR\",\n",
    "        \"NRAS\": \"RAS\",\n",
    "        \"KRAS\": \"RAS\",\n",
    "        \"HRAS\": \"RAS\",\n",
    "        \"AKT2\": \"AKT\",\n",
    "        \"MYCN\": \"MYC\",\n",
    "        }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ef4a4e44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'analysis/Refametinib/genes_proteins_models/MEK1_MEK2_target_block/models/generic/Montagud2022_Prostate_Cancer_original.cfg'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "subdir = f\"{'_'.join(drug_targets)}_target_block\"\n",
    "folder_generic_models = f\"analysis/{drug_interest}/{type_model}/{subdir}/models/generic/\"\n",
    "folder_models = f\"analysis/{drug_interest}/{type_model}/{subdir}/models\"\n",
    "patients_categ = ['resistant', 'sensitive', 'healthy']\n",
    "if not os.path.exists(f\"analysis/{drug_interest}/{type_model}/{subdir}/models/generic\"):\n",
    "        os.makedirs(f\"analysis/{drug_interest}/{type_model}/{subdir}/models/generic\")\n",
    "\n",
    "for patient_categ in patients_categ:\n",
    "    if not os.path.exists(f\"analysis/{drug_interest}/{type_model}/{subdir}/results/{patient_categ}\"):\n",
    "        os.makedirs(f\"analysis/{drug_interest}/{type_model}/{subdir}/results/{patient_categ}\")\n",
    "\n",
    "    if not os.path.exists(f\"analysis/{drug_interest}/{type_model}/{subdir}/models/{patient_categ}\"):\n",
    "        os.makedirs(f\"analysis/{drug_interest}/{type_model}/{subdir}/models/{patient_categ}\")\n",
    "\n",
    "dest_dir = f\"analysis/{drug_interest}/{type_model}/{subdir}/models/generic\"\n",
    "os.makedirs(dest_dir, exist_ok=True)\n",
    "\n",
    "# Copy the files\n",
    "shutil.copy('data/montagud_models/Montagud2022_Prostate_Cancer_original.bnd', dest_dir)\n",
    "shutil.copy('data/montagud_models/Montagud2022_Prostate_Cancer_original.cfg', dest_dir) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08c0a0b3",
   "metadata": {},
   "source": [
    "### Pre-processing the data (genes, proteins, cnv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f2854a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/romane/repos/icr/Attractor-Resistance/functions/generate_utils/identification_patients/get_patients_sens_res.py:56: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  annotations_models_filtered.rename(\n",
      "/Users/romane/repos/icr/Attractor-Resistance/functions/generate_utils/pre_process_data/pre_process_genes.py:83: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  rna_seq_data.rename(columns={\"gene_symbol\": \"gene_name\"}, inplace=True)\n",
      "/Users/romane/repos/icr/Attractor-Resistance/functions/generate_utils/pre_process_data/pre_process_proteins.py:145: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  proteins_data_patient_id_filtered[\"symbol\"] = proteins_data_patient_id_filtered[\n",
      "/Users/romane/repos/icr/Attractor-Resistance/functions/generate_utils/pre_process_data/pre_process_proteins.py:107: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  proteins_data.rename(columns={\"protein_symbol\": \"protein_name\"}, inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "preprocessed  proteins\n"
     ]
    }
   ],
   "source": [
    "top_resistant_ids, top_sensitive_ids, top_healthy_ids, montagud_nodes, rna_seq_data_filtered, cnv_data_filtered, table_rna_seq_patients, df_melted_protein, table_proteins_patients = pre_process_re(\n",
    "montagud_data,\n",
    "rna_seq_data,\n",
    "cnv_data,\n",
    "number_patients,\n",
    "drug_data,\n",
    "annotations_models,\n",
    "drug_interest,\n",
    "proteins_data,\n",
    "type_model,\n",
    "name_montagud_maps,\n",
    "nodes_to_add,\n",
    "synonyms_maps,\n",
    "tissue_interest=None,\n",
    "tissue_remove=None,\n",
    "nodes_to_remove = node_to_remove,\n",
    ")\n",
    "\n",
    "os.makedirs(f\"analysis/{drug_interest}/data_filtered\", exist_ok=True)\n",
    "rna_seq_data_filtered.to_csv(f\"analysis/{drug_interest}/data_filtered/rna_seq_data_filtered.csv\")\n",
    "cnv_data_filtered.to_csv(f\"analysis/{drug_interest}/data_filtered/cnv_data_filtered.csv\")\n",
    "table_rna_seq_patients.to_csv(f\"analysis/{drug_interest}/data_filtered/table_rna_seq_patients.csv\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "08a49a86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'AR_ERG' in montagud_nodes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a59784de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All .cfg and .bnd files created for sensitive, resistant and healthy patients.\n",
      "🔍 Processing patient SIDM00415, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00415: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00415, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00415: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00623, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00623: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00623, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00623: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00793, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00793: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00793, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00793: CNV — nodes modified\n",
      "🔍 Processing patient SIDM01018, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM01018: CNV — nodes modified\n",
      "🔍 Processing patient SIDM01018, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM01018: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00340, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00340: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00340, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00340: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00346, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00346: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00346, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00346: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00417, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00417: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00417, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00417: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00886, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00886: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00886, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00886: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00859, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00859: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00859, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00859: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00391, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00391: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00391, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00391: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00338, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00338: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00338, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00338: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00549, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00549: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00549, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00549: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00413, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00413: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00413, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00413: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00438, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00438: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00438, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00438: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00448, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00448: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00448, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00448: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00459, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00459: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00459, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00459: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00407, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00407: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00407, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00407: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00437, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00437: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00437, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00437: CNV — nodes modified\n",
      "🔍 Processing patient SIDM01023, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM01023: CNV — nodes modified\n",
      "🔍 Processing patient SIDM01023, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM01023: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00277, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00277: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00277, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00277: CNV — nodes modified\n",
      "🔍 Processing patient SIDM01005, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM01005: CNV — nodes modified\n",
      "🔍 Processing patient SIDM01005, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM01005: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00233, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00233: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00233, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00233: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00657, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00657: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00657, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00657: CNV — nodes modified\n",
      "🔍 Processing patient SIDM01015, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM01015: CNV — nodes modified\n",
      "🔍 Processing patient SIDM01015, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM01015: CNV — nodes modified\n",
      "🔍 Processing patient SIDM01017, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM01017: CNV — nodes modified\n",
      "🔍 Processing patient SIDM01017, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM01017: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00412, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00412: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00412, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00412: CNV — nodes modified\n",
      "🔍 Processing patient SIDM01038, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM01038: CNV — nodes modified\n",
      "🔍 Processing patient SIDM01038, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM01038: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00406, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00406: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00406, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00406: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00561, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00561: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00561, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00561: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00576, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00576: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00576, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00576: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00813, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00813: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00813, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00813: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00072, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00072: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00072, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00072: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00632, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00632: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00632, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00632: CNV — nodes modified\n",
      "🔍 Processing patient SIDM01049, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM01049: CNV — nodes modified\n",
      "🔍 Processing patient SIDM01049, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM01049: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00900, gene: CHK1\n",
      "CHK1 node found. Replacing...\n",
      "SIDM00900: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00900, gene: CHK2\n",
      "CHK2 node found. Replacing...\n",
      "SIDM00900: CNV — nodes modified\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00417_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00793_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00340_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00346_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM01018_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00338_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00859_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00413_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00623_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00886_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00448_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00438_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00415_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00391_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00549_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00886_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00417_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00448_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00438_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00338_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00859_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00793_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00413_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00340_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00623_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00391_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00549_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00415_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00346_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM01018_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00338_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00859_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00793_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00340_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00413_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00623_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00417_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00886_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00448_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00438_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00346_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00415_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM01018_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00391_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/resistant/pers_models/SIDM00549_AZD7762.cfg\n",
      "🔍 Processing patient SIDM00415, gene: SMO\n",
      "Patient SIDM00415 is in both gain and loss groups. Please review.\n",
      "SMO node found. Replacing...\n",
      "🔍 Processing patient SIDM00415, gene: EZH2\n",
      "Patient SIDM00415 is in both gain and loss groups. Please review.\n",
      "EZH2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00415, gene: BCL2\n",
      "Patient SIDM00415 is in both gain and loss groups. Please review.\n",
      "BCL2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00415, gene: NCOR1\n",
      "Patient SIDM00415 is in both gain and loss groups. Please review.\n",
      "NCOR1 node found. Replacing...\n",
      "SIDM00415: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00793, gene: SMO\n",
      "SMO node found. Replacing...\n",
      "🔍 Processing patient SIDM00793, gene: EZH2\n",
      "EZH2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00793, gene: ETV1\n",
      "ETV1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00793, gene: EGFR\n",
      "EGFR node found. Replacing...\n",
      "SIDM00793: CNV — nodes modified\n",
      "No .bnd file found for patient: SIDM00459\n",
      "No .bnd file found for patient: SIDM01023\n",
      "🔍 Processing patient SIDM00346, gene: ATM\n",
      "ATM node found. Replacing...\n",
      "🔍 Processing patient SIDM00346, gene: BRCA1\n",
      "BRCA1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00346, gene: CCND1\n",
      "CCND1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00346, gene: DAXX\n",
      "DAXX node found. Replacing...\n",
      "🔍 Processing patient SIDM00346, gene: ERG\n",
      "ERG node found. Replacing...\n",
      "🔍 Processing patient SIDM00346, gene: TSC1\n",
      "TSC1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00346, gene: TERT\n",
      "TERT node found. Replacing...\n",
      "🔍 Processing patient SIDM00346, gene: SPOP\n",
      "SPOP node found. Replacing...\n",
      "🔍 Processing patient SIDM00346, gene: NF1\n",
      "NF1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00346, gene: MYC\n",
      "MYC node found. Replacing...\n",
      "🔍 Processing patient SIDM00346, gene: MAP3K1\n",
      "MAP3K1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00346, gene: CASP8\n",
      "CASP8 node found. Replacing...\n",
      "🔍 Processing patient SIDM00346, gene: AR\n",
      "AR node found. Replacing...\n",
      "🔍 Processing patient SIDM00346, gene: NCOR1\n",
      "NCOR1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00346, gene: IDH1\n",
      "IDH1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00346, gene: EGFR\n",
      "EGFR node found. Replacing...\n",
      "🔍 Processing patient SIDM00346, gene: BCL2\n",
      "BCL2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00346, gene: ETV1\n",
      "ETV1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00346, gene: EZH2\n",
      "EZH2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00346, gene: MED12\n",
      "MED12 node found. Replacing...\n",
      "🔍 Processing patient SIDM00346, gene: PTCH1\n",
      "PTCH1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00346, gene: SMO\n",
      "SMO node found. Replacing...\n",
      "SIDM00346: CNV — nodes modified\n",
      "No .bnd file found for patient: SIDM00233\n",
      "No .bnd file found for patient: SIDM01017\n",
      "No .bnd file found for patient: SIDM00412\n",
      "No .bnd file found for patient: SIDM00406\n",
      "No .bnd file found for patient: SIDM00072\n",
      "🔍 Processing patient SIDM00340, gene: ATR\n",
      "Patient SIDM00340 is in both gain and loss groups. Please review.\n",
      "ATR node found. Replacing...\n",
      "🔍 Processing patient SIDM00340, gene: CASP8\n",
      "Patient SIDM00340 is in both gain and loss groups. Please review.\n",
      "CASP8 node found. Replacing...\n",
      "🔍 Processing patient SIDM00340, gene: TSC1\n",
      "Patient SIDM00340 is in both gain and loss groups. Please review.\n",
      "TSC1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00340, gene: NF1\n",
      "Patient SIDM00340 is in both gain and loss groups. Please review.\n",
      "NF1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00340, gene: EZH2\n",
      "Patient SIDM00340 is in both gain and loss groups. Please review.\n",
      "EZH2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00340, gene: TSC1\n",
      "Patient SIDM00340 is in both gain and loss groups. Please review.\n",
      "TSC1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00340, gene: BRCA2\n",
      "Patient SIDM00340 is in both gain and loss groups. Please review.\n",
      "BRCA2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00340, gene: FOXA1\n",
      "Patient SIDM00340 is in both gain and loss groups. Please review.\n",
      "FOXA1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00340, gene: NF1\n",
      "Patient SIDM00340 is in both gain and loss groups. Please review.\n",
      "NF1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00340, gene: RB1\n",
      "Patient SIDM00340 is in both gain and loss groups. Please review.\n",
      "RB1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00340, gene: RB1\n",
      "Patient SIDM00340 is in both gain and loss groups. Please review.\n",
      "RB1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00340, gene: BRCA2\n",
      "Patient SIDM00340 is in both gain and loss groups. Please review.\n",
      "BRCA2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00340, gene: IDH1\n",
      "Patient SIDM00340 is in both gain and loss groups. Please review.\n",
      "IDH1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00340, gene: ATR\n",
      "Patient SIDM00340 is in both gain and loss groups. Please review.\n",
      "ATR node found. Replacing...\n",
      "🔍 Processing patient SIDM00340, gene: DAXX\n",
      "Patient SIDM00340 is in both gain and loss groups. Please review.\n",
      "DAXX node found. Replacing...\n",
      "🔍 Processing patient SIDM00340, gene: MYC\n",
      "Patient SIDM00340 is in both gain and loss groups. Please review.\n",
      "MYC node found. Replacing...\n",
      "🔍 Processing patient SIDM00340, gene: MYC\n",
      "Patient SIDM00340 is in both gain and loss groups. Please review.\n",
      "MYC node found. Replacing...\n",
      "🔍 Processing patient SIDM00340, gene: FOXA1\n",
      "Patient SIDM00340 is in both gain and loss groups. Please review.\n",
      "FOXA1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00340, gene: ETV1\n",
      "Patient SIDM00340 is in both gain and loss groups. Please review.\n",
      "ETV1 node found. Replacing...\n",
      "SIDM00340: CNV — nodes modified\n",
      "No .bnd file found for patient: SIDM00277\n",
      "No .bnd file found for patient: SIDM01005\n",
      "🔍 Processing patient SIDM00417, gene: ATM\n",
      "Patient SIDM00417 is in both gain and loss groups. Please review.\n",
      "ATM node found. Replacing...\n",
      "🔍 Processing patient SIDM00417, gene: BRCA2\n",
      "Patient SIDM00417 is in both gain and loss groups. Please review.\n",
      "BRCA2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00417, gene: BAX\n",
      "Patient SIDM00417 is in both gain and loss groups. Please review.\n",
      "BAX node found. Replacing...\n",
      "🔍 Processing patient SIDM00417, gene: ERG\n",
      "Patient SIDM00417 is in both gain and loss groups. Please review.\n",
      "ERG node found. Replacing...\n",
      "🔍 Processing patient SIDM00417, gene: ETV1\n",
      "Patient SIDM00417 is in both gain and loss groups. Please review.\n",
      "ETV1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00417, gene: FGFR3\n",
      "Patient SIDM00417 is in both gain and loss groups. Please review.\n",
      "FGFR3 node found. Replacing...\n",
      "🔍 Processing patient SIDM00417, gene: SMO\n",
      "Patient SIDM00417 is in both gain and loss groups. Please review.\n",
      "SMO node found. Replacing...\n",
      "🔍 Processing patient SIDM00417, gene: SPOP\n",
      "Patient SIDM00417 is in both gain and loss groups. Please review.\n",
      "SPOP node found. Replacing...\n",
      "🔍 Processing patient SIDM00417, gene: RB1\n",
      "Patient SIDM00417 is in both gain and loss groups. Please review.\n",
      "RB1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00417, gene: NCOR1\n",
      "Patient SIDM00417 is in both gain and loss groups. Please review.\n",
      "NCOR1 node found. Replacing...\n",
      "SIDM00417: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00413, gene: PTEN\n",
      "Patient SIDM00413 is in both gain and loss groups. Please review.\n",
      "PTEN node found. Replacing...\n",
      "🔍 Processing patient SIDM00413, gene: EGFR\n",
      "Patient SIDM00413 is in both gain and loss groups. Please review.\n",
      "EGFR node found. Replacing...\n",
      "🔍 Processing patient SIDM00413, gene: EGFR\n",
      "Patient SIDM00413 is in both gain and loss groups. Please review.\n",
      "EGFR node found. Replacing...\n",
      "🔍 Processing patient SIDM00413, gene: ETV1\n",
      "Patient SIDM00413 is in both gain and loss groups. Please review.\n",
      "ETV1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00413, gene: MYC\n",
      "Patient SIDM00413 is in both gain and loss groups. Please review.\n",
      "MYC node found. Replacing...\n",
      "🔍 Processing patient SIDM00413, gene: ETV1\n",
      "Patient SIDM00413 is in both gain and loss groups. Please review.\n",
      "ETV1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00413, gene: PTEN\n",
      "Patient SIDM00413 is in both gain and loss groups. Please review.\n",
      "PTEN node found. Replacing...\n",
      "🔍 Processing patient SIDM00413, gene: RB1\n",
      "Patient SIDM00413 is in both gain and loss groups. Please review.\n",
      "RB1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00413, gene: RB1\n",
      "Patient SIDM00413 is in both gain and loss groups. Please review.\n",
      "RB1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00413, gene: MYC\n",
      "Patient SIDM00413 is in both gain and loss groups. Please review.\n",
      "MYC node found. Replacing...\n",
      "SIDM00413: CNV — nodes modified\n",
      "No .bnd file found for patient: SIDM00437\n",
      "🔍 Processing patient SIDM00886, gene: SMO\n",
      "SMO node found. Replacing...\n",
      "🔍 Processing patient SIDM00886, gene: MYC\n",
      "MYC node found. Replacing...\n",
      "🔍 Processing patient SIDM00886, gene: EZH2\n",
      "EZH2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00886, gene: ETV1\n",
      "ETV1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00886, gene: EGFR\n",
      "EGFR node found. Replacing...\n",
      "🔍 Processing patient SIDM00886, gene: FGFR3\n",
      "FGFR3 node found. Replacing...\n",
      "🔍 Processing patient SIDM00886, gene: PTEN\n",
      "PTEN node found. Replacing...\n",
      "🔍 Processing patient SIDM00886, gene: PTCH1\n",
      "PTCH1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00886, gene: VHL\n",
      "VHL node found. Replacing...\n",
      "🔍 Processing patient SIDM00886, gene: CTNNB1\n",
      "CTNNB1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00886, gene: TSC1\n",
      "TSC1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00886, gene: RB1\n",
      "RB1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00886, gene: BRCA2\n",
      "BRCA2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00886, gene: ATR\n",
      "ATR node found. Replacing...\n",
      "SIDM00886: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00391, gene: BCL2\n",
      "Patient SIDM00391 is in both gain and loss groups. Please review.\n",
      "BCL2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00391, gene: EZH2\n",
      "Patient SIDM00391 is in both gain and loss groups. Please review.\n",
      "EZH2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00391, gene: BRCA2\n",
      "Patient SIDM00391 is in both gain and loss groups. Please review.\n",
      "BRCA2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00391, gene: CDH1\n",
      "Patient SIDM00391 is in both gain and loss groups. Please review.\n",
      "CDH1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00391, gene: JUN\n",
      "Patient SIDM00391 is in both gain and loss groups. Please review.\n",
      "JUN node found. Replacing...\n",
      "🔍 Processing patient SIDM00391, gene: FOXA1\n",
      "Patient SIDM00391 is in both gain and loss groups. Please review.\n",
      "FOXA1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00391, gene: RB1\n",
      "Patient SIDM00391 is in both gain and loss groups. Please review.\n",
      "RB1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00391, gene: CCND1\n",
      "Patient SIDM00391 is in both gain and loss groups. Please review.\n",
      "CCND1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00391, gene: ATM\n",
      "Patient SIDM00391 is in both gain and loss groups. Please review.\n",
      "ATM node found. Replacing...\n",
      "🔍 Processing patient SIDM00391, gene: SMO\n",
      "Patient SIDM00391 is in both gain and loss groups. Please review.\n",
      "SMO node found. Replacing...\n",
      "SIDM00391: CNV — nodes modified\n",
      "No .bnd file found for patient: SIDM01015\n",
      "No .bnd file found for patient: SIDM00900\n",
      "No .bnd file found for patient: SIDM01038\n",
      "🔍 Processing patient SIDM00438, gene: RB1\n",
      "RB1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00438, gene: BRCA2\n",
      "BRCA2 node found. Replacing...\n",
      "SIDM00438: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00448, gene: BRCA2\n",
      "BRCA2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00448, gene: EGFR\n",
      "EGFR node found. Replacing...\n",
      "🔍 Processing patient SIDM00448, gene: ETV1\n",
      "ETV1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00448, gene: EZH2\n",
      "EZH2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00448, gene: MYC\n",
      "MYC node found. Replacing...\n",
      "🔍 Processing patient SIDM00448, gene: RB1\n",
      "RB1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00448, gene: SMO\n",
      "SMO node found. Replacing...\n",
      "🔍 Processing patient SIDM00448, gene: TERT\n",
      "TERT node found. Replacing...\n",
      "🔍 Processing patient SIDM00448, gene: TSC1\n",
      "TSC1 node found. Replacing...\n",
      "SIDM00448: CNV — nodes modified\n",
      "No .bnd file found for patient: SIDM00576\n",
      "No .bnd file found for patient: SIDM00407\n",
      "🔍 Processing patient SIDM01018, gene: AXIN1\n",
      "Patient SIDM01018 is in both gain and loss groups. Please review.\n",
      "AXIN1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01018, gene: PTEN\n",
      "Patient SIDM01018 is in both gain and loss groups. Please review.\n",
      "PTEN node found. Replacing...\n",
      "🔍 Processing patient SIDM01018, gene: CDH1\n",
      "Patient SIDM01018 is in both gain and loss groups. Please review.\n",
      "CDH1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01018, gene: EZH2\n",
      "Patient SIDM01018 is in both gain and loss groups. Please review.\n",
      "EZH2 node found. Replacing...\n",
      "🔍 Processing patient SIDM01018, gene: ATM\n",
      "Patient SIDM01018 is in both gain and loss groups. Please review.\n",
      "ATM node found. Replacing...\n",
      "🔍 Processing patient SIDM01018, gene: BCL2\n",
      "Patient SIDM01018 is in both gain and loss groups. Please review.\n",
      "BCL2 node found. Replacing...\n",
      "🔍 Processing patient SIDM01018, gene: BRCA2\n",
      "Patient SIDM01018 is in both gain and loss groups. Please review.\n",
      "BRCA2 node found. Replacing...\n",
      "🔍 Processing patient SIDM01018, gene: CCND1\n",
      "Patient SIDM01018 is in both gain and loss groups. Please review.\n",
      "CCND1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01018, gene: FOXA1\n",
      "Patient SIDM01018 is in both gain and loss groups. Please review.\n",
      "FOXA1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01018, gene: ETV1\n",
      "Patient SIDM01018 is in both gain and loss groups. Please review.\n",
      "ETV1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01018, gene: EP300\n",
      "Patient SIDM01018 is in both gain and loss groups. Please review.\n",
      "EP300 node found. Replacing...\n",
      "🔍 Processing patient SIDM01018, gene: MDM2\n",
      "Patient SIDM01018 is in both gain and loss groups. Please review.\n",
      "MDM2 node found. Replacing...\n",
      "🔍 Processing patient SIDM01018, gene: RB1\n",
      "Patient SIDM01018 is in both gain and loss groups. Please review.\n",
      "RB1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01018, gene: SMO\n",
      "Patient SIDM01018 is in both gain and loss groups. Please review.\n",
      "SMO node found. Replacing...\n",
      "🔍 Processing patient SIDM01018, gene: TERT\n",
      "Patient SIDM01018 is in both gain and loss groups. Please review.\n",
      "TERT node found. Replacing...\n",
      "🔍 Processing patient SIDM01018, gene: TSC2\n",
      "Patient SIDM01018 is in both gain and loss groups. Please review.\n",
      "TSC2 node found. Replacing...\n",
      "SIDM01018: CNV — nodes modified\n",
      "No .bnd file found for patient: SIDM00657\n",
      "🔍 Processing patient SIDM00338, gene: CDH1\n",
      "Patient SIDM00338 is in both gain and loss groups. Please review.\n",
      "CDH1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00338, gene: CASP8\n",
      "Patient SIDM00338 is in both gain and loss groups. Please review.\n",
      "CASP8 node found. Replacing...\n",
      "🔍 Processing patient SIDM00338, gene: IDH1\n",
      "Patient SIDM00338 is in both gain and loss groups. Please review.\n",
      "IDH1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00338, gene: ATM\n",
      "Patient SIDM00338 is in both gain and loss groups. Please review.\n",
      "ATM node found. Replacing...\n",
      "🔍 Processing patient SIDM00338, gene: TSC1\n",
      "Patient SIDM00338 is in both gain and loss groups. Please review.\n",
      "TSC1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00338, gene: CCND1\n",
      "Patient SIDM00338 is in both gain and loss groups. Please review.\n",
      "CCND1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00338, gene: ATM\n",
      "Patient SIDM00338 is in both gain and loss groups. Please review.\n",
      "ATM node found. Replacing...\n",
      "🔍 Processing patient SIDM00338, gene: PTCH1\n",
      "Patient SIDM00338 is in both gain and loss groups. Please review.\n",
      "PTCH1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00338, gene: BCL2\n",
      "Patient SIDM00338 is in both gain and loss groups. Please review.\n",
      "BCL2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00338, gene: BCL2\n",
      "Patient SIDM00338 is in both gain and loss groups. Please review.\n",
      "BCL2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00338, gene: NCOR1\n",
      "Patient SIDM00338 is in both gain and loss groups. Please review.\n",
      "NCOR1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00338, gene: CDH1\n",
      "Patient SIDM00338 is in both gain and loss groups. Please review.\n",
      "CDH1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00338, gene: IDH1\n",
      "Patient SIDM00338 is in both gain and loss groups. Please review.\n",
      "IDH1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00338, gene: NCOR1\n",
      "Patient SIDM00338 is in both gain and loss groups. Please review.\n",
      "NCOR1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00338, gene: PTCH1\n",
      "Patient SIDM00338 is in both gain and loss groups. Please review.\n",
      "PTCH1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00338, gene: TSC1\n",
      "Patient SIDM00338 is in both gain and loss groups. Please review.\n",
      "TSC1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00338, gene: CASP8\n",
      "Patient SIDM00338 is in both gain and loss groups. Please review.\n",
      "CASP8 node found. Replacing...\n",
      "🔍 Processing patient SIDM00338, gene: MYC\n",
      "Patient SIDM00338 is in both gain and loss groups. Please review.\n",
      "MYC node found. Replacing...\n",
      "🔍 Processing patient SIDM00338, gene: MYC\n",
      "Patient SIDM00338 is in both gain and loss groups. Please review.\n",
      "MYC node found. Replacing...\n",
      "🔍 Processing patient SIDM00338, gene: CCND1\n",
      "Patient SIDM00338 is in both gain and loss groups. Please review.\n",
      "CCND1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00338, gene: ETV1\n",
      "Patient SIDM00338 is in both gain and loss groups. Please review.\n",
      "ETV1 node found. Replacing...\n",
      "SIDM00338: CNV — nodes modified\n",
      "No .bnd file found for patient: SIDM00561\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00657_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00459_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00407_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00412_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM01017_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM01023_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00576_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM01038_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00406_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00561_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00233_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM01005_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00437_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM01015_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00277_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00561_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00657_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00233_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00459_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM01038_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00407_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00412_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00406_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM01017_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00437_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM01015_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00277_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM01023_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM01005_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00576_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00459_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM01038_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00407_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00412_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM01017_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00406_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00561_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00657_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00233_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM01023_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00576_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM01005_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00437_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM01015_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/sensitive/pers_models/SIDM00277_AZD7762.cfg\n",
      "No .bnd file found for patient: SIDM00415\n",
      "No .bnd file found for patient: SIDM00793\n",
      "🔍 Processing patient SIDM00459, gene: BCL2\n",
      "BCL2 node found. Replacing...\n",
      "SIDM00459: CNV — nodes modified\n",
      "🔍 Processing patient SIDM01023, gene: RB1\n",
      "Patient SIDM01023 is in both gain and loss groups. Please review.\n",
      "RB1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01023, gene: BRCA2\n",
      "Patient SIDM01023 is in both gain and loss groups. Please review.\n",
      "BRCA2 node found. Replacing...\n",
      "🔍 Processing patient SIDM01023, gene: CDH1\n",
      "Patient SIDM01023 is in both gain and loss groups. Please review.\n",
      "CDH1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01023, gene: DAXX\n",
      "Patient SIDM01023 is in both gain and loss groups. Please review.\n",
      "DAXX node found. Replacing...\n",
      "🔍 Processing patient SIDM01023, gene: EP300\n",
      "Patient SIDM01023 is in both gain and loss groups. Please review.\n",
      "EP300 node found. Replacing...\n",
      "🔍 Processing patient SIDM01023, gene: MDM2\n",
      "Patient SIDM01023 is in both gain and loss groups. Please review.\n",
      "MDM2 node found. Replacing...\n",
      "🔍 Processing patient SIDM01023, gene: NCOR1\n",
      "Patient SIDM01023 is in both gain and loss groups. Please review.\n",
      "NCOR1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01023, gene: SPOP\n",
      "Patient SIDM01023 is in both gain and loss groups. Please review.\n",
      "SPOP node found. Replacing...\n",
      "🔍 Processing patient SIDM01023, gene: ERG\n",
      "Patient SIDM01023 is in both gain and loss groups. Please review.\n",
      "ERG node found. Replacing...\n",
      "SIDM01023: CNV — nodes modified\n",
      "No .bnd file found for patient: SIDM00346\n",
      "🔍 Processing patient SIDM00233, gene: RB1\n",
      "Patient SIDM00233 is in both gain and loss groups. Please review.\n",
      "RB1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00233, gene: ATM\n",
      "Patient SIDM00233 is in both gain and loss groups. Please review.\n",
      "ATM node found. Replacing...\n",
      "🔍 Processing patient SIDM00233, gene: CTNNB1\n",
      "Patient SIDM00233 is in both gain and loss groups. Please review.\n",
      "CTNNB1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00233, gene: VHL\n",
      "Patient SIDM00233 is in both gain and loss groups. Please review.\n",
      "VHL node found. Replacing...\n",
      "🔍 Processing patient SIDM00233, gene: TERT\n",
      "Patient SIDM00233 is in both gain and loss groups. Please review.\n",
      "TERT node found. Replacing...\n",
      "🔍 Processing patient SIDM00233, gene: SMO\n",
      "Patient SIDM00233 is in both gain and loss groups. Please review.\n",
      "SMO node found. Replacing...\n",
      "🔍 Processing patient SIDM00233, gene: FGFR3\n",
      "Patient SIDM00233 is in both gain and loss groups. Please review.\n",
      "FGFR3 node found. Replacing...\n",
      "🔍 Processing patient SIDM00233, gene: EZH2\n",
      "Patient SIDM00233 is in both gain and loss groups. Please review.\n",
      "EZH2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00233, gene: ERG\n",
      "Patient SIDM00233 is in both gain and loss groups. Please review.\n",
      "ERG node found. Replacing...\n",
      "SIDM00233: CNV — nodes modified\n",
      "🔍 Processing patient SIDM01017, gene: EGFR\n",
      "Patient SIDM01017 is in both gain and loss groups. Please review.\n",
      "EGFR node found. Replacing...\n",
      "🔍 Processing patient SIDM01017, gene: MYC\n",
      "Patient SIDM01017 is in both gain and loss groups. Please review.\n",
      "MYC node found. Replacing...\n",
      "🔍 Processing patient SIDM01017, gene: NCOR1\n",
      "Patient SIDM01017 is in both gain and loss groups. Please review.\n",
      "NCOR1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01017, gene: SMO\n",
      "Patient SIDM01017 is in both gain and loss groups. Please review.\n",
      "SMO node found. Replacing...\n",
      "🔍 Processing patient SIDM01017, gene: EZH2\n",
      "Patient SIDM01017 is in both gain and loss groups. Please review.\n",
      "EZH2 node found. Replacing...\n",
      "SIDM01017: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00412, gene: BRCA1\n",
      "Patient SIDM00412 is in both gain and loss groups. Please review.\n",
      "BRCA1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00412, gene: BRCA2\n",
      "Patient SIDM00412 is in both gain and loss groups. Please review.\n",
      "BRCA2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00412, gene: SPOP\n",
      "Patient SIDM00412 is in both gain and loss groups. Please review.\n",
      "SPOP node found. Replacing...\n",
      "🔍 Processing patient SIDM00412, gene: RB1\n",
      "Patient SIDM00412 is in both gain and loss groups. Please review.\n",
      "RB1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00412, gene: NF1\n",
      "Patient SIDM00412 is in both gain and loss groups. Please review.\n",
      "NF1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00412, gene: CDH1\n",
      "Patient SIDM00412 is in both gain and loss groups. Please review.\n",
      "CDH1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00412, gene: FGFR3\n",
      "Patient SIDM00412 is in both gain and loss groups. Please review.\n",
      "FGFR3 node found. Replacing...\n",
      "🔍 Processing patient SIDM00412, gene: NCOR1\n",
      "Patient SIDM00412 is in both gain and loss groups. Please review.\n",
      "NCOR1 node found. Replacing...\n",
      "SIDM00412: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00406, gene: EZH2\n",
      "Patient SIDM00406 is in both gain and loss groups. Please review.\n",
      "EZH2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00406, gene: MED12\n",
      "Patient SIDM00406 is in both gain and loss groups. Please review.\n",
      "MED12 node found. Replacing...\n",
      "🔍 Processing patient SIDM00406, gene: CASP8\n",
      "Patient SIDM00406 is in both gain and loss groups. Please review.\n",
      "CASP8 node found. Replacing...\n",
      "🔍 Processing patient SIDM00406, gene: BRCA1\n",
      "Patient SIDM00406 is in both gain and loss groups. Please review.\n",
      "BRCA1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00406, gene: ETV1\n",
      "Patient SIDM00406 is in both gain and loss groups. Please review.\n",
      "ETV1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00406, gene: AR\n",
      "Patient SIDM00406 is in both gain and loss groups. Please review.\n",
      "AR node found. Replacing...\n",
      "🔍 Processing patient SIDM00406, gene: IDH1\n",
      "Patient SIDM00406 is in both gain and loss groups. Please review.\n",
      "IDH1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00406, gene: DAXX\n",
      "Patient SIDM00406 is in both gain and loss groups. Please review.\n",
      "DAXX node found. Replacing...\n",
      "🔍 Processing patient SIDM00406, gene: CCND1\n",
      "Patient SIDM00406 is in both gain and loss groups. Please review.\n",
      "CCND1 node found. Replacing...\n",
      "SIDM00406: CNV — nodes modified\n",
      "No .bnd file found for patient: SIDM00072\n",
      "No .bnd file found for patient: SIDM00340\n",
      "🔍 Processing patient SIDM00277, gene: PTCH1\n",
      "Patient SIDM00277 is in both gain and loss groups. Please review.\n",
      "PTCH1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00277, gene: EZH2\n",
      "Patient SIDM00277 is in both gain and loss groups. Please review.\n",
      "EZH2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00277, gene: FGFR3\n",
      "Patient SIDM00277 is in both gain and loss groups. Please review.\n",
      "FGFR3 node found. Replacing...\n",
      "🔍 Processing patient SIDM00277, gene: MYC\n",
      "Patient SIDM00277 is in both gain and loss groups. Please review.\n",
      "MYC node found. Replacing...\n",
      "🔍 Processing patient SIDM00277, gene: TSC1\n",
      "Patient SIDM00277 is in both gain and loss groups. Please review.\n",
      "TSC1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00277, gene: DAXX\n",
      "Patient SIDM00277 is in both gain and loss groups. Please review.\n",
      "DAXX node found. Replacing...\n",
      "🔍 Processing patient SIDM00277, gene: ATM\n",
      "Patient SIDM00277 is in both gain and loss groups. Please review.\n",
      "ATM node found. Replacing...\n",
      "🔍 Processing patient SIDM00277, gene: BAX\n",
      "Patient SIDM00277 is in both gain and loss groups. Please review.\n",
      "BAX node found. Replacing...\n",
      "SIDM00277: CNV — nodes modified\n",
      "🔍 Processing patient SIDM01005, gene: SPOP\n",
      "Patient SIDM01005 is in both gain and loss groups. Please review.\n",
      "SPOP node found. Replacing...\n",
      "🔍 Processing patient SIDM01005, gene: PTEN\n",
      "Patient SIDM01005 is in both gain and loss groups. Please review.\n",
      "PTEN node found. Replacing...\n",
      "🔍 Processing patient SIDM01005, gene: MYC\n",
      "Patient SIDM01005 is in both gain and loss groups. Please review.\n",
      "MYC node found. Replacing...\n",
      "🔍 Processing patient SIDM01005, gene: TSC2\n",
      "Patient SIDM01005 is in both gain and loss groups. Please review.\n",
      "TSC2 node found. Replacing...\n",
      "🔍 Processing patient SIDM01005, gene: AXIN1\n",
      "Patient SIDM01005 is in both gain and loss groups. Please review.\n",
      "AXIN1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01005, gene: BRCA2\n",
      "Patient SIDM01005 is in both gain and loss groups. Please review.\n",
      "BRCA2 node found. Replacing...\n",
      "🔍 Processing patient SIDM01005, gene: IDH1\n",
      "Patient SIDM01005 is in both gain and loss groups. Please review.\n",
      "IDH1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01005, gene: NCOR1\n",
      "Patient SIDM01005 is in both gain and loss groups. Please review.\n",
      "NCOR1 node found. Replacing...\n",
      "SIDM01005: CNV — nodes modified\n",
      "No .bnd file found for patient: SIDM00417\n",
      "No .bnd file found for patient: SIDM00413\n",
      "🔍 Processing patient SIDM00437, gene: MYC\n",
      "Patient SIDM00437 is in both gain and loss groups. Please review.\n",
      "MYC node found. Replacing...\n",
      "🔍 Processing patient SIDM00437, gene: RB1\n",
      "Patient SIDM00437 is in both gain and loss groups. Please review.\n",
      "RB1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00437, gene: BAX\n",
      "Patient SIDM00437 is in both gain and loss groups. Please review.\n",
      "BAX node found. Replacing...\n",
      "🔍 Processing patient SIDM00437, gene: BRCA2\n",
      "Patient SIDM00437 is in both gain and loss groups. Please review.\n",
      "BRCA2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00437, gene: DAXX\n",
      "Patient SIDM00437 is in both gain and loss groups. Please review.\n",
      "DAXX node found. Replacing...\n",
      "🔍 Processing patient SIDM00437, gene: RB1\n",
      "Patient SIDM00437 is in both gain and loss groups. Please review.\n",
      "RB1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00437, gene: MYC\n",
      "Patient SIDM00437 is in both gain and loss groups. Please review.\n",
      "MYC node found. Replacing...\n",
      "🔍 Processing patient SIDM00437, gene: BRCA2\n",
      "Patient SIDM00437 is in both gain and loss groups. Please review.\n",
      "BRCA2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00437, gene: BAX\n",
      "Patient SIDM00437 is in both gain and loss groups. Please review.\n",
      "BAX node found. Replacing...\n",
      "🔍 Processing patient SIDM00437, gene: DAXX\n",
      "Patient SIDM00437 is in both gain and loss groups. Please review.\n",
      "DAXX node found. Replacing...\n",
      "🔍 Processing patient SIDM00437, gene: ETV1\n",
      "Patient SIDM00437 is in both gain and loss groups. Please review.\n",
      "ETV1 node found. Replacing...\n",
      "SIDM00437: CNV — nodes modified\n",
      "No .bnd file found for patient: SIDM00886\n",
      "No .bnd file found for patient: SIDM00391\n",
      "🔍 Processing patient SIDM01015, gene: NCOR1\n",
      "Patient SIDM01015 is in both gain and loss groups. Please review.\n",
      "NCOR1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01015, gene: MYC\n",
      "Patient SIDM01015 is in both gain and loss groups. Please review.\n",
      "MYC node found. Replacing...\n",
      "🔍 Processing patient SIDM01015, gene: EZH2\n",
      "Patient SIDM01015 is in both gain and loss groups. Please review.\n",
      "EZH2 node found. Replacing...\n",
      "🔍 Processing patient SIDM01015, gene: NCOR1\n",
      "Patient SIDM01015 is in both gain and loss groups. Please review.\n",
      "NCOR1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01015, gene: PTCH1\n",
      "Patient SIDM01015 is in both gain and loss groups. Please review.\n",
      "PTCH1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01015, gene: PTCH1\n",
      "Patient SIDM01015 is in both gain and loss groups. Please review.\n",
      "PTCH1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01015, gene: ETV1\n",
      "Patient SIDM01015 is in both gain and loss groups. Please review.\n",
      "ETV1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01015, gene: MYC\n",
      "Patient SIDM01015 is in both gain and loss groups. Please review.\n",
      "MYC node found. Replacing...\n",
      "🔍 Processing patient SIDM01015, gene: BCL2\n",
      "Patient SIDM01015 is in both gain and loss groups. Please review.\n",
      "BCL2 node found. Replacing...\n",
      "🔍 Processing patient SIDM01015, gene: EZH2\n",
      "Patient SIDM01015 is in both gain and loss groups. Please review.\n",
      "EZH2 node found. Replacing...\n",
      "🔍 Processing patient SIDM01015, gene: BCL2\n",
      "Patient SIDM01015 is in both gain and loss groups. Please review.\n",
      "BCL2 node found. Replacing...\n",
      "SIDM01015: CNV — nodes modified\n",
      "No .bnd file found for patient: SIDM00900\n",
      "🔍 Processing patient SIDM01038, gene: PTCH1\n",
      "Patient SIDM01038 is in both gain and loss groups. Please review.\n",
      "PTCH1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01038, gene: EP300\n",
      "Patient SIDM01038 is in both gain and loss groups. Please review.\n",
      "EP300 node found. Replacing...\n",
      "🔍 Processing patient SIDM01038, gene: EZH2\n",
      "Patient SIDM01038 is in both gain and loss groups. Please review.\n",
      "EZH2 node found. Replacing...\n",
      "🔍 Processing patient SIDM01038, gene: NF1\n",
      "Patient SIDM01038 is in both gain and loss groups. Please review.\n",
      "NF1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01038, gene: BRCA1\n",
      "Patient SIDM01038 is in both gain and loss groups. Please review.\n",
      "BRCA1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01038, gene: ATM\n",
      "Patient SIDM01038 is in both gain and loss groups. Please review.\n",
      "ATM node found. Replacing...\n",
      "🔍 Processing patient SIDM01038, gene: AXIN1\n",
      "Patient SIDM01038 is in both gain and loss groups. Please review.\n",
      "AXIN1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01038, gene: BAX\n",
      "Patient SIDM01038 is in both gain and loss groups. Please review.\n",
      "BAX node found. Replacing...\n",
      "🔍 Processing patient SIDM01038, gene: JUN\n",
      "Patient SIDM01038 is in both gain and loss groups. Please review.\n",
      "JUN node found. Replacing...\n",
      "🔍 Processing patient SIDM01038, gene: MDM2\n",
      "Patient SIDM01038 is in both gain and loss groups. Please review.\n",
      "MDM2 node found. Replacing...\n",
      "🔍 Processing patient SIDM01038, gene: NCOR1\n",
      "Patient SIDM01038 is in both gain and loss groups. Please review.\n",
      "NCOR1 node found. Replacing...\n",
      "🔍 Processing patient SIDM01038, gene: SMO\n",
      "Patient SIDM01038 is in both gain and loss groups. Please review.\n",
      "SMO node found. Replacing...\n",
      "🔍 Processing patient SIDM01038, gene: TERT\n",
      "Patient SIDM01038 is in both gain and loss groups. Please review.\n",
      "TERT node found. Replacing...\n",
      "SIDM01038: CNV — nodes modified\n",
      "No .bnd file found for patient: SIDM00438\n",
      "No .bnd file found for patient: SIDM00448\n",
      "🔍 Processing patient SIDM00576, gene: SMO\n",
      "Patient SIDM00576 is in both gain and loss groups. Please review.\n",
      "SMO node found. Replacing...\n",
      "🔍 Processing patient SIDM00576, gene: EZH2\n",
      "Patient SIDM00576 is in both gain and loss groups. Please review.\n",
      "EZH2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00576, gene: FGFR3\n",
      "Patient SIDM00576 is in both gain and loss groups. Please review.\n",
      "FGFR3 node found. Replacing...\n",
      "🔍 Processing patient SIDM00576, gene: MYC\n",
      "Patient SIDM00576 is in both gain and loss groups. Please review.\n",
      "MYC node found. Replacing...\n",
      "🔍 Processing patient SIDM00576, gene: BCL2\n",
      "Patient SIDM00576 is in both gain and loss groups. Please review.\n",
      "BCL2 node found. Replacing...\n",
      "SIDM00576: CNV — nodes modified\n",
      "🔍 Processing patient SIDM00407, gene: MYC\n",
      "Patient SIDM00407 is in both gain and loss groups. Please review.\n",
      "MYC node found. Replacing...\n",
      "🔍 Processing patient SIDM00407, gene: MYC\n",
      "Patient SIDM00407 is in both gain and loss groups. Please review.\n",
      "MYC node found. Replacing...\n",
      "🔍 Processing patient SIDM00407, gene: DAXX\n",
      "Patient SIDM00407 is in both gain and loss groups. Please review.\n",
      "DAXX node found. Replacing...\n",
      "🔍 Processing patient SIDM00407, gene: EGFR\n",
      "Patient SIDM00407 is in both gain and loss groups. Please review.\n",
      "EGFR node found. Replacing...\n",
      "🔍 Processing patient SIDM00407, gene: ETV1\n",
      "Patient SIDM00407 is in both gain and loss groups. Please review.\n",
      "ETV1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00407, gene: DAXX\n",
      "Patient SIDM00407 is in both gain and loss groups. Please review.\n",
      "DAXX node found. Replacing...\n",
      "🔍 Processing patient SIDM00407, gene: EGFR\n",
      "Patient SIDM00407 is in both gain and loss groups. Please review.\n",
      "EGFR node found. Replacing...\n",
      "🔍 Processing patient SIDM00407, gene: ETV1\n",
      "Patient SIDM00407 is in both gain and loss groups. Please review.\n",
      "ETV1 node found. Replacing...\n",
      "SIDM00407: CNV — nodes modified\n",
      "No .bnd file found for patient: SIDM01018\n",
      "🔍 Processing patient SIDM00657, gene: MYC\n",
      "MYC node found. Replacing...\n",
      "🔍 Processing patient SIDM00657, gene: BAX\n",
      "BAX node found. Replacing...\n",
      "SIDM00657: CNV — nodes modified\n",
      "No .bnd file found for patient: SIDM00338\n",
      "🔍 Processing patient SIDM00561, gene: SMO\n",
      "Patient SIDM00561 is in both gain and loss groups. Please review.\n",
      "SMO node found. Replacing...\n",
      "🔍 Processing patient SIDM00561, gene: VHL\n",
      "Patient SIDM00561 is in both gain and loss groups. Please review.\n",
      "VHL node found. Replacing...\n",
      "🔍 Processing patient SIDM00561, gene: AR\n",
      "Patient SIDM00561 is in both gain and loss groups. Please review.\n",
      "AR node found. Replacing...\n",
      "🔍 Processing patient SIDM00561, gene: EZH2\n",
      "Patient SIDM00561 is in both gain and loss groups. Please review.\n",
      "EZH2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00561, gene: FGFR3\n",
      "Patient SIDM00561 is in both gain and loss groups. Please review.\n",
      "FGFR3 node found. Replacing...\n",
      "🔍 Processing patient SIDM00561, gene: MED12\n",
      "Patient SIDM00561 is in both gain and loss groups. Please review.\n",
      "MED12 node found. Replacing...\n",
      "🔍 Processing patient SIDM00561, gene: NCOR1\n",
      "Patient SIDM00561 is in both gain and loss groups. Please review.\n",
      "NCOR1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00561, gene: BCL2\n",
      "Patient SIDM00561 is in both gain and loss groups. Please review.\n",
      "BCL2 node found. Replacing...\n",
      "🔍 Processing patient SIDM00561, gene: ATM\n",
      "Patient SIDM00561 is in both gain and loss groups. Please review.\n",
      "ATM node found. Replacing...\n",
      "🔍 Processing patient SIDM00561, gene: MDM2\n",
      "Patient SIDM00561 is in both gain and loss groups. Please review.\n",
      "MDM2 node found. Replacing...\n",
      "SIDM00561: CNV — nodes modified\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/healthy/pers_models/SIDM00632_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/healthy/pers_models/SIDM00072_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/healthy/pers_models/SIDM00813_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/healthy/pers_models/SIDM01049_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/healthy/pers_models/SIDM00900_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/healthy/pers_models/SIDM00632_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/healthy/pers_models/SIDM00072_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/healthy/pers_models/SIDM00813_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/healthy/pers_models/SIDM01049_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/healthy/pers_models/SIDM00900_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/healthy/pers_models/SIDM00813_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/healthy/pers_models/SIDM01049_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/healthy/pers_models/SIDM00900_AZD7762.cfg\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/healthy/pers_models/SIDM00632_AZD7762.bnd\n",
      "Modified and saved: analysis/blood_tissue/AZD7762/genes_models/CHK1_CHK2_target_block/models/healthy/pers_models/SIDM00072_AZD7762.bnd\n",
      "No .bnd file found for patient: SIDM00415\n",
      "No .bnd file found for patient: SIDM00793\n",
      "No .bnd file found for patient: SIDM00459\n",
      "No .bnd file found for patient: SIDM01023\n",
      "No .bnd file found for patient: SIDM00346\n",
      "No .bnd file found for patient: SIDM00233\n",
      "No .bnd file found for patient: SIDM01017\n",
      "No .bnd file found for patient: SIDM00412\n",
      "No .bnd file found for patient: SIDM00406\n",
      "🔍 Processing patient SIDM00072, gene: FOXA1\n",
      "FOXA1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00072, gene: DAXX\n",
      "DAXX node found. Replacing...\n",
      "SIDM00072: CNV — nodes modified\n",
      "No .bnd file found for patient: SIDM00340\n",
      "No .bnd file found for patient: SIDM00277\n",
      "No .bnd file found for patient: SIDM01005\n",
      "No .bnd file found for patient: SIDM00417\n",
      "No .bnd file found for patient: SIDM00413\n",
      "No .bnd file found for patient: SIDM00437\n",
      "No .bnd file found for patient: SIDM00886\n",
      "No .bnd file found for patient: SIDM00391\n",
      "No .bnd file found for patient: SIDM01015\n",
      "🔍 Processing patient SIDM00900, gene: NCOR1\n",
      "Patient SIDM00900 is in both gain and loss groups. Please review.\n",
      "NCOR1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00900, gene: CTNNB1\n",
      "Patient SIDM00900 is in both gain and loss groups. Please review.\n",
      "CTNNB1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00900, gene: MYC\n",
      "Patient SIDM00900 is in both gain and loss groups. Please review.\n",
      "MYC node found. Replacing...\n",
      "🔍 Processing patient SIDM00900, gene: PTCH1\n",
      "Patient SIDM00900 is in both gain and loss groups. Please review.\n",
      "PTCH1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00900, gene: TSC1\n",
      "Patient SIDM00900 is in both gain and loss groups. Please review.\n",
      "TSC1 node found. Replacing...\n",
      "🔍 Processing patient SIDM00900, gene: TERT\n",
      "Patient SIDM00900 is in both gain and loss groups. Please review.\n",
      "TERT node found. Replacing...\n",
      "🔍 Processing patient SIDM00900, gene: CDH1\n",
      "Patient SIDM00900 is in both gain and loss groups. Please review.\n",
      "CDH1 node found. Replacing...\n",
      "SIDM00900: CNV — nodes modified\n",
      "No .bnd file found for patient: SIDM01038\n",
      "No .bnd file found for patient: SIDM00438\n",
      "No .bnd file found for patient: SIDM00448\n",
      "No .bnd file found for patient: SIDM00576\n",
      "No .bnd file found for patient: SIDM00407\n",
      "No .bnd file found for patient: SIDM01018\n",
      "No .bnd file found for patient: SIDM00657\n",
      "No .bnd file found for patient: SIDM00338\n",
      "No .bnd file found for patient: SIDM00561\n"
     ]
    }
   ],
   "source": [
    "generate_models_re(\n",
    "            folder_generic_models,\n",
    "            folder_models,\n",
    "            top_resistant_ids,\n",
    "            top_sensitive_ids,\n",
    "            top_healthy_ids,\n",
    "            drug_interest,\n",
    "            drug_targets,\n",
    "            phenotype_interest,\n",
    "            rna_seq_data,\n",
    "            montagud_nodes,\n",
    "            table_rna_seq_patients,\n",
    "            cnv_data_filtered,\n",
    "            name_montagud_maps,\n",
    "            type_model,\n",
    "            df_melted_protein,\n",
    "            table_proteins_patients,\n",
    "            nodes_to_remove,\n",
    "            intervention_gene = None,\n",
    "        )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "500a80dd",
   "metadata": {},
   "source": [
    "### Create the personalized PAN cancer models (genes and cnv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "af280270",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "set()\n"
     ]
    }
   ],
   "source": [
    "# check no intersection between the two groups\n",
    "intersection = set(top_resistant_ids) & set(top_sensitive_ids) & set(top_healthy_ids)\n",
    "print(intersection)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d5b22ac",
   "metadata": {},
   "source": [
    "### All pipeline (create personalized networks and compute attractors distribution)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "191b77e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ALL PIPELINE \n",
    "\n",
    "\n",
    "# Pre process data and saved them in the data filtered folder\n",
    "# AS tissue interest is PAN cancer -> tissue_interest = None (want all tissues)\n",
    "\n",
    "\n",
    "drugs_dict = {'Refametinib': ['MEK1', 'MEK2'], 'Pictilisib': ['PI3K'], 'AZD7762': ['CHK1', 'CHK2']}\n",
    "type_models = ['proteins_models', 'genes_models', 'genes_proteins_models']\n",
    "for drug in drugs_dict:\n",
    "    for type_model in type_models:\n",
    "        subdir = f\"{'_'.join(drugs_dict[drug])}_target_block\"\n",
    "        folder_generic_models = f\"analysis/{drug}/{type_model}/{subdir}/models/generic/\"\n",
    "        folder_models = f\"analysis/{drug}/{type_model}/{subdir}/models\"\n",
    "        folder_results = f\"analysis/{drug}/{type_model}/{subdir}\"\n",
    "\n",
    "\n",
    "        patients_categ = ['resistant', 'sensitive', 'healthy']\n",
    "        \n",
    "\n",
    "        for patient_categ in patients_categ:\n",
    "            os.makedirs(f\"analysis/{drug}/{type_model}/{subdir}/results/{patient_categ}\", exist_ok=True)\n",
    "            os.makedirs(f\"analysis/{drug}/{type_model}/{subdir}/models/{patient_categ}\", exist_ok=True)\n",
    "\n",
    "        dest_dir = f\"analysis/{drug}/{type_model}/{subdir}/models/generic\"\n",
    "        os.makedirs(dest_dir, exist_ok=True)\n",
    "\n",
    "        # Copy the files\n",
    "        shutil.copy('data/montagud_models/Montagud2022_Prostate_Cancer_original.bnd', dest_dir)\n",
    "        shutil.copy('data/montagud_models/Montagud2022_Prostate_Cancer_original.cfg', dest_dir)\n",
    "\n",
    "        top_resistant_ids, top_sensitive_ids, top_healthy_ids, montagud_nodes, rna_seq_data_filtered, cnv_data_filtered, table_rna_seq_patients, df_melted_protein, table_proteins_patients = pre_process_re(\n",
    "            montagud_data.copy(),\n",
    "            rna_seq_data.copy(),\n",
    "            cnv_data.copy(),\n",
    "            number_patients,\n",
    "            drug_data.copy(),\n",
    "            annotations_models.copy(),\n",
    "            drug,\n",
    "            proteins_data.copy(),\n",
    "            type_model,\n",
    "            name_montagud_maps,\n",
    "            nodes_to_add,\n",
    "            synonyms_maps,\n",
    "            tissue_interest=None,\n",
    "            tissue_remove=None,\n",
    "            node_to_remove=None,\n",
    "        )\n",
    "\n",
    "        # os.makedirs(f\"analysis/{drug}/data_filtered\", exist_ok=True)\n",
    "        # rna_seq_data_filtered.to_csv(f\"analysis/{drug}/data_filtered/rna_seq_data_filtered.csv\")\n",
    "        # cnv_data_filtered.to_csv(f\"analysis/{drug}/data_filtered/cnv_data_filtered.csv\")\n",
    "        # table_rna_seq_patients.to_csv(f\"analysis/{drug}/data_filtered/table_rna_seq_patients.csv\")\n",
    "\n",
    "\n",
    "\n",
    "# rna_seq_data not filtered as get the may expression value\n",
    "        # generate_models_re(\n",
    "        #     folder_generic_models,\n",
    "        #     folder_models,\n",
    "        #     top_resistant_ids,\n",
    "        #     top_sensitive_ids,\n",
    "        #     top_healthy_ids,\n",
    "        #     drug,\n",
    "        #     drugs_dict[drug],\n",
    "        #     phenotype_interest,\n",
    "        #     rna_seq_data,\n",
    "        #     montagud_nodes,\n",
    "        #     table_rna_seq_patients,\n",
    "        #     cnv_data_filtered,\n",
    "        #     name_montagud_maps,\n",
    "        #     type_model,\n",
    "        #     df_melted_protein,\n",
    "        #     table_proteins_patients,\n",
    "        #     intervention_gene = None,\n",
    "        # )\n",
    "\n",
    "        nb_patients_required = downstream_analysis(folder_results,folder_models, drug, top_resistant_ids, top_sensitive_ids, top_healthy_ids, patients_categ, inputs_list, phenotype_interest, annotations_models, list_active_inputs = None)\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
